---
title: "`r paste(params$sample_id, str_replace_all(params$report_name, '[\\._-]', ' '))`"
author: "`r paste('StemCNV-check, version:',  params$version)`"
output:
  html_document:
    df_print: paged
    toc: true
    toc_float: true
  pdf_document:
    default
params:
  sample_id: ''
  inputs: NULL
  config: NULL
  report_config: NULL
  config_delta: NULL
  report_name: 'StemCNV-check-report'
  out_format: 'html'
  workdir: ''
  basedir: ''
  gtf_file: ''
  ginfo_file: ''
  dosage_file: ''
  version: 'unversioned'
  verbosity: 0
  log_dir: ''
---

<style type="text/css">
.main-container {
  max-width: 100% !important;
  margin: auto;
}
.badge-red {
  background-color:#ff0000;
  box-sizing: border-box;
}
.badge-orange {
  background-color:#ffa500;
  box-sizing: border-box;
}
.nav-pills {
  # visibility: hidden;
  display: none;
}
[role=tab] {
  height: 40px;
  width: 250px;
}
# collapsible text boxes
details {
  user-select: none;
}
details>summary span.icon {
  width: 24px;
  height: 24px;
  transition: all 0.3s;
  margin-left: auto;
}
details[open] summary span.icon {
  transform: rotate(180deg);
}
summary {
  display: flex;
  cursor: pointer;
  background-color:#eee;
  color: #444;
}
summary::-webkit-details-marker {
  display: none;
}

</style>


```{r setup, include=FALSE}
library(plyranges)
library(tidyverse)
library(dendextend)
library(ggpubr)
library(scales)
library(RIdeogram)
library(DT)
library(yaml)
library(patchwork)
library(knitr)
library(kableExtra)
library(sessioninfo)
library(readxl)
`%!in%` <- Negate(`%in%`)

config <- params$config
report_config <- params$report_config 

# Load helper functions
source(file.path(config$snakedir, 'scripts/R/helper_functions.R'))
source(file.path(config$snakedir, 'scripts/R/hotspot_functions.R'))
source(file.path(config$snakedir, 'scripts/R/report_table_functions.R'))
source(file.path(config$snakedir, 'scripts/R/report_plotting_functions.R'))
source(file.path(config$snakedir, 'scripts/R/vcf_io_functions.R'))

# General variables & settings

image_folder <- paste0(params$report_name, '-', params$out_format, '_images/')
workdir <- ifelse(params$workdir == '', getwd(), params$workdir)
knitr::opts_chunk$set(
    echo = FALSE, message=NA, warning=NA, error=FALSE,
    fig.path = file.path(workdir, image_folder)
)
options(knitr.kable.NA = '')

defined_labels <- get_defined_labels(config)

sampletable <- config$sample_table %>%
    fix_rel_filepath(config) %>%
    read_sampletable()

sample_id <- params$sample_id
sex <- get_sample_info(sample_id, 'sex', config, sampletable)
ref_id <- get_sample_info(sample_id, 'ref_id', config, sampletable)
sex.ref <- get_sample_info(ref_id, 'sex', config, sampletable)
array <- sampletable %>%
    filter(Sample_ID == sample_id) %>%
    pull(Array_Name)

if (!is.na(ref_id)) {
    sample_headers <- c(sample_id, paste0('Reference (', ref_id, ')'))
    names(sample_headers) <- c(sample_id, ref_id)
    use_ids <- c(sample_id, ref_id)
} else {
    sample_headers <- sample_id
    names(sample_headers) <- sample_id
    use_ids <- sample_id
}

target_chrom_style <- get_target_chrom_style(config, parse_cnv_vcf(params$input$cnv_vcf))

roi_tb <- get_roi_tb(sample_id, sampletable, config)
stemcell_hotspot_tb <- load_hotspot_table(config, 'stemcell_hotspot')
dosage_sensitive_gene_tb <- get_dosage_sensivity_tb(
    snakemake@params$dosage_file,
    config
)
cancer_gene_tb <- load_hotspot_table(config, 'cancer_gene')
snv_hotspot_tb <- load_hotspot_table(config, 'snv_hotspot')

hotspot_genes <- bind_rows(
    stemcell_hotspot_tb,
    cancer_gene_tb,
    dosage_sensitive_gene_tb,
    roi_tb
) %>%
    filter(mapping == 'gene_name') %>%
    pull(hotspot)
hotspot_genes <- c(
    hotspot_genes,
    snv_hotspot_tb$gene_name
) %>% unique()

gr_genes <- load_gtf_data(params$gtf_file, config, target_chrom_style, include_hotspot_genes = hotspot_genes)
gr_info  <- load_genomeInfo(params$ginfo_file, config, target_chrom_style)

roi_gr <- get_roi_gr(roi_tb, gr_genes, gr_info, target_chrom_style)

## general functions

tr_tibble <- function(tb) {
    tr <- t(tb )
    if (nrow(tb) == 1) {
        colnames(tr) <- sample_id
    } else {
        colnames(tr) <- sample_headers[tr[1,]]
    }
    tb <- as_tibble(tr, rownames = ' ')
    tb[2:nrow(tb),]
}

## Rmd related functions


include.section <- function(section) {
    include <- ifelse(
        all(report_config$include_sections == '__all__'), 
        T,
        section %in% report_config$include_sections
    )
    exclude <- section %in% report_config$exclude_sections
    # also exclude QC sections if they are not defined in settings
    include <- include & !(
        str_detect(section, '^QC\\.(?!(summary|GenCall|settings))') &
        str_remove(section, 'QC\\.') %!in% config$settings$CNV.calling.tools
    )
    include & !exclude
}

include_roi_plots <-  include.section('regions.of.interest') & ifelse(
    'Regions_of_Interest' %in% colnames(sampletable),
    length(roi_gr) > 0,
    FALSE
)

CNV_ID_str <- function(ID, i, region_name = NULL, ...) {
    #ROIs might have user defined names
    if(!is.null(region_name)) id_str <- paste0('roi-', region_name)
    # All other calls
    else id_str <- str_glue('nr{i}-{ID}')
    str_to_lower(id_str) %>%
        str_replace_all('[:]', '')
}


make_CNV_plot_section <- function(call.table, plotsection = 'denovo') {
    if (nrow(call.table) == 0) {
        return(str_glue('No {plotsection} CNV calls found.'))
    }
    always_include <- report_config$call.data.and.plots[[plotsection]]$always_include_CNVs
    for (i in call.table$i) {

        row <- call.table[i,]
        res <- make_call_plot(
            row, raw_LRR_BAF, cnv_calls, gr_genes, gr_info, sample_headers, defined_labels,
            report_config$call.data.and.plots[[plotsection]]$plot.region.minsize,
            report_config$call.data.and.plots[[plotsection]]$plot.flanking.region.relative
        )
        if(i <= report_config$call.data.and.plots[[plotsection]]$min_number_plots | row$Call_label %in% always_include) {
            cat('### ', pmap_chr(row, CNV_ID_str), '\n\n')
            if (typeof(res$gg) == 'character') {
                cat(res$gg, '\n\n')
                next()
            }
            if (report_config$call.data.and.plots[[plotsection]]$include.plot) {
                subchunkify(res$gg, str_glue('CNV_call.{plotsection}.nr{i}.plot'), 5, 10)
                cat('\n\n')
            }
            # message(str_glue('N hotspots: {length(res$hotspots)}'))
            if (report_config$call.data.and.plots[[plotsection]]$include.hotspot.table & length(res$hotspots)>0) {
                res$hotspots %>%
                    hotspot_table_output(
                        as.character(row$CNV_type), plotsection,
                        stemcell_hotspot_tb, dosage_sensitive_gene_tb, cancer_gene_tb,
                        report_config, params$out_format
                    ) %>%
                    subchunkify(str_glue('CNV_call.{plotsection}.nr{i}.hotspots'))
                cat('\n\n')
            }
            if (report_config$call.data.and.plots[[plotsection]]$include.gene.table.details != 'None') {
                gene_table_output(res$genes, plotsection, stemcell_hotspot_tb, dosage_sensitive_gene_tb, cancer_gene_tb, report_config) %>%
                    subchunkify(str_glue('CNV_call.{plotsection}.nr{i}.table.genes'))
                cat('\n\n')
            }
            cat('\n\n')
        } else if (typeof(res$gg) != 'character') {
            ggsave(
                file.path(image_folder, str_glue('CNV_call.{plotsection}.nr{i}.plot-1.png')),
                res$gg,
                width = 20, height = 10, dpi=200
            )
        }
    }
}

# Make a separate subchunnk, this is often needed when mixing loops, tabsets & html elements
subchunkify <- function(g, id, fig_height=7, fig_width=7, options = '') {
  g_deparsed <- paste0(deparse(function() {g}), collapse = '')

  sub_chunk <- paste0("```{r ", id, ", fig.height=", fig_height, ", fig.width=", fig_width,
                      ifelse(options == '', options, paste(',', options)), ", echo=FALSE}\n",
  "(",  g_deparsed, ")()",
  "\n```")

  cat(knitr::knit(text = knitr::knit_expand(text = sub_chunk), quiet = TRUE))
}

collapsible_html_text <- function(text, header = 'Description' ) {
    paste0(
        '<details>\n\n<summary>\n\n', header, '\n\n',
        '<span class="icon">â‡©</span>\n\n',
        '</summary>\n\n',
        '<p>\n\n',
        text,
        '</p>\n\n',
        '</details>\n\n'
    ) %>% cat()
}

# Additional RData file for interactive report testing
if (params$verbosity >= 3) {
    save.image(file = file.path(, 'R_env_report.Rdata'))
}

```


# Sample Overview

```{r sample.information, eval = include.section('sample.information'), results = 'asis'}

cat('## Sample Information\n\n')

info.tb <- sampletable %>%
  filter(Sample_ID %in% c(sample_id, ref_id)) %>%
  dplyr::select(Sample_ID, Sex, Reference_Sample, Array_Name, any_of(report_config$sample.info.extra.cols)) %>%
  mutate(`Analysis run date` = Sys.Date()) %>%
  tr_tibble() %>%
  dplyr::select(-starts_with('Reference ('))

simple_table_output(info.tb, params$out_format)

cat('\n\n')

```

## QC metrics {.tabset}

```{r qc.summary, eval = include.section('QC.summary'), results='asis'}
cat('### Summary\n\n')

crit_excl_filters <- ifelse(
    is.null(config$evaluation_settings$CNV_call_labels$`Critical de-novo`$not_allowed_vcf_filters),
    '<none>',
    config$evaluation_settings$CNV_call_labels$`Critical de-novo`$not_allowed_vcf_filters %>% paste(collapse = ', ')
)
repo_excl_filters <- ifelse(
    is.null(config$evaluation_settings$CNV_call_labels$`Reportable de-novo`$not_allowed_vcf_filters),
    '<none>',
    config$evaluation_settings$CNV_call_labels$`Reportable de-novo`$not_allowed_vcf_filters %>% paste(collapse = ', ')
)

paste0(
    'These summary tables are meant to serve as a quick overview of the quality of an hPSC sample. ',
    'The first table contains QC metrics primarily related to data quality (affected by both the DNA used and the array ',
    'run itself), this table will also display values from the reference sample if possible. The second table contains ',
    'QC metrics related to the potentially problematic CNVs and SNVs identified in only the primary sample. ',
    'Note that without a reference sample all detected variants are considered "de-novo", even if most are likely from ',
    'the original donor germline.\n\n',
    'Coloring of all fields is based on (usually) two thresholds defined in the config file (under the evaluation_settings section), ',
    'one for a signalling (yellow) level, and one for a more serious warning (orange) or even critical (red) level. ',
    'Only certain values are potentially considered critical and are marked by bold text in the table, ',
    'which values behave like this is also defined in the config.'
) %>% collapsible_html_text('Data QC explanations')

summary_tables <- read_excel(
    fix_rel_filepath(params$input$summary_xlsx, config),
    sheet = 'summary_stats'
) %>%
    summary_table(sample_headers, config, defined_labels, params$out_format)

summary_tables[[1]]
toolconfig <- config$settings$CNV_processing$call_processing

paste0(
    'For assigning SNVs (as determined by the Chip Array) as critical or reportable, the SNV genotype needs to differ ',
    'from the reference sample (without a reference SNV can not be distinguished from gerline mutations) ',
    'and it needs to match specific SNV categories:\n\n',
    'Critical SNVs:\n\n', 
    paste(sapply(
        config$settings$SNV_analysis$critical_SNV, \(x) str_glue(' - {x}: ', defined_labels$SNV_categories[[x]])
        ), collapse = '\n\n'
    ),
    '\n\nReportable SNVs:\n\n',
    paste(sapply(
        config$settings$SNV_analysis$reportable_SNV, \(x) str_glue(' - {x}: ', defined_labels$SNV_categories[[x]])
        ), collapse = '\n\n'
    ),
    '\n\n',
    'All CNV calls are given a single label designation which takes into account a minimum Check_Score threshold, ',
    'overlap with a reference call and certain call filter flags (i.e. calls below minimum size, or having a probe ',
    'coverage gap). The defined call labels are:\n\n',
    paste(imap(
        config$evaluation_settings$CNV_call_labels, \(x, name) {
            str_glue('
                 - {name}:
                   Minimum required Check-Score: {x$minimum_check_score}
                   Exclusion of calls with any filter among: {paste(x$not_allowed_vcf_filters, collapse=", ")}
                   Match to CNV in reference sample: {ifelse(x$reference_match, "required", "not allowed")}'                
            ) %>% str_replace_all('\\n', '\n\n')
        }), collapse = '\n\n'
    ),
    #Add explanations for each CNV filter from labels def file
    '\n\nThe defined CNV filter flags are:\n\n',
    paste(imap(
        defined_labels$vcf_filters, \(x, name) str_glue('- {name}: ', x)
        ), collapse = '\n\n'
    )
) %>% collapsible_html_text('Sample QC explanations')

summary_tables[[2]]

cat('\n\n')
```

```{r qc.gencall, eval = include.section('QC.GenCall'), results='asis'}
cat('### GenCall\n\n')

cat(
    'This table displays the direct quality metrics from the GenCall software.\n\n'
)

read_excel(
    fix_rel_filepath(params$input$summary_xlsx, config),
    sheet = 'gencall_stats'
) %>%
    simple_table_output(params$out_format)
cat('\n\n')
```

```{r qc.penncnv, eval = include.section('QC.PennCNV'), results='asis'}
cat('### PennCNV\n\n')
cat(
    'The first table displays quality metrics from the PennCNV algorythm. ',
    'The second displays CNV call statistcs for only PennCNV\n\n'
)

read_excel(
    fix_rel_filepath(params$input$summary_xlsx, config),
    sheet = 'PennCNV_QC_info'
) %>%
    simple_table_output(params$out_format)

read_excel(
    fix_rel_filepath(params$input$summary_xlsx, config),
    sheet = 'PennCNV_stats'
) %>%
    simple_table_output(params$out_format)
cat('\n\n')
```

```{r qc.CBS, eval = include.section('QC.CBS'), results='asis'}
cat('### CBS\n\n')
cat(
    'This table displays CNV call statistcs for only CBS.\n\n'
)

read_excel(
    fix_rel_filepath(params$input$summary_xlsx, config),
    sheet = 'CBS_stats'
) %>%
    simple_table_output(params$out_format)
cat('\n\n')
```

```{r config, eval = include.section('QC.settings'), results='asis'}
cat('### Config\n\n')
config_delta <- params$config_delta %>%
    lapply(\(sublist) {
        names(sublist) <- names(sublist) %>%
            str_replace("root\\['", "config: ") %>%
            str_replace_all("'\\]\\['", " : ") %>%
            str_remove("'\\]") %>%
            str_replace('report_settings', paste('reports :', params$report_name, ': '))
        sublist
     })

cat('Changes from default config:\n\n')
cat(knitr::knit(text = knitr::knit_expand(text = '```yaml\n`r as.yaml(config_delta)`\n```'), quiet = TRUE))
cat('\n\n')

cat('Complete config used by StemCNV-check:\n\n')
cat(knitr::knit(text = knitr::knit_expand(text = '```yaml\n`r as.yaml(config)`\n```'), quiet = TRUE))
cat('\n\n')
# session_info()
# cat('\n\n')
cat('### R session info\n\n')
```

```{r sessioninfo, eval = include.section('QC.settings')}
session_info()
```

# CNV calling {.tabset}

```{r CNV.call.plot.data}
raw_LRR_BAF <- c(params$input$snp_vcf, params$input$ref_snp_vcf) %>%
    sapply(fix_rel_filepath, config=config) %>%
    lapply(parse_snp_vcf, apply_filter = FALSE) %>%
    bind_ranges() %>%
    fix_CHROM_format(target_chrom_style) %>%
    as_tibble() %>%
    #FIXME / future: adapt this
    dplyr::rename(Chr = seqnames, Position = start,
                  `B Allele Freq` = BAF, `Log R Ratio` = LRR) %>%
    mutate(filter.passed = FILTER == 'PASS') %>%
    dplyr::select(sample_id, ID, Chr, Position, filter.passed, `Log R Ratio`, `B Allele Freq`) 

cnv_calls <- c(params$input$cnv_vcf, params$input$ref_cnv_vcf) %>%
    sapply(fix_rel_filepath, config=config) %>%
    lapply(parse_cnv_vcf) %>%
    bind_ranges() %>%
    fix_CHROM_format(target_chrom_style) %>%
    as_tibble() %>%
	filter(Call_label %!in% report_config$CNV_call_labels_removed) %>%
    mutate(
        CNV_type = factor(CNV_type),
        reference_overlap = factor(!is.na(reference_coverage)),
        Call_label = factor(Call_label, levels = names(config$evaluation_settings$CNV_call_labels)),
        FILTER = ifelse(is.na(FILTER), '', FILTER),
        probe_coverage_gap = str_detect(FILTER, 'probe_gap'),
        high_probe_density = str_detect(FILTER, 'high_probe_dens'),
        caller_merging_coverage = sapply(initial_call_details, \(x){
            if (is.na(x)) return(NA)
            str_split(x, '\\|') %>%
                unlist() %>%
                str_remove('[0-9]+-[0-9]+_CN[0-9]_cov') %>%
                str_replace('_[^_]+$', '%') %>%
                str_replace('_', ':') %>%
                paste(collapse = '; ')
        }),
    ) 

#for comparison/overlap factor levels need to include all chromosomes and have same order
cnv_calls$seqnames <- fct_expand(cnv_calls$seqnames, levels(raw_LRR_BAF$Chr))
cnv_calls$seqnames <- fct_relevel(cnv_calls$seqnames, levels(raw_LRR_BAF$Chr))

denovo_calls.table <- cnv_calls %>% 
    dplyr::rename(chrom = seqnames, Size = width) %>%
    arrange(Call_label, desc(Check_Score), desc(Size), chrom, start) %>%
    filter(sample_id == !!sample_id & Call_label %in% report_config$call.data.and.plots$denovo$call_labels_include) %>%
    mutate(i = row_number())

reference_calls.table <- cnv_calls %>% 
    dplyr::rename(chrom = seqnames, Size = width) %>%
    arrange(desc(Check_Score), desc(Size), chrom, start) %>%
    filter(sample_id == !!sample_id & Call_label %in% report_config$call.data.and.plots$reference_gt$call_labels_include) %>%
    mutate(i = row_number())

```

```{r denovo_calls.header, eval = include.section('denovo_calls.table') | include.section('denovo_calls.plots'), results='asis'}
cat('## de-novo CNV calls {.tabset .tabset-pills}\n\n')
paste(
    'This section describes all de-novo CNV calls, meaning calls without a match in the reference sample.',
    'The table allows sorting and filtering the calls by various criteria, default is sorting by Check-Score.',
    'Hovering over the column headers gives explanations for each column and the "Column visibility" button can be',
    'used to change the default selection of visible columns.\n\n',
    'The section immediately below the table contains details for each CNV call, including a plot of the CNV region,',
    '(if relevant) a table of annotated genes and hotspots, and a table of all genes overlapping the CNV (or plot region).\n\n'
) %>% collapsible_html_text('de-novo CNV calls table')
```

```{r denovo_calls.table, results='asis', eval = include.section('denovo_calls.table')}
CNV_table_output(
    denovo_calls.table, 'denovo',
    stemcell_hotspot_tb, dosage_sensitive_gene_tb, cancer_gene_tb,
    gr_info, report_config#, caption = 'de-novo CNV calls'
)
cat('\n\n')
```

```{r denovo_calls.plots, eval = include.section('denovo_calls.plots'), results='asis'}
make_CNV_plot_section(denovo_calls.table, 'denovo')
cat('\n\n')
```
```{r reference_gt_calls.header, eval = !is.na(ref_id) & (include.section('reference_gt_calls.table') | include.section('reference_gt_calls.plots')), results='asis'}
cat('## Reference genotype CNV calls {.tabset .tabset-pills}\n\n')
paste(
    'This section describes all reference CNV calls, meaning calls for which a match in the reference sample was found.',
    str_glue(
        'Matching of CNV calls is based on a minimum of at least ',
        '{config$settings$CNV_processing$call_processing$min.reciprocal.coverage.with.ref}% ',
        'reciprocal overlap between sample and reference.\n\n'
    ),
    'Otherwise this section uses the same layout as to the de-novo calls section.\n\n'
) %>% collapsible_html_text('reference genotype CNV calls table')
```

```{r reference_gt_calls.table, results='asis', eval = !is.na(ref_id) & include.section('reference_gt_calls.table')}
CNV_table_output(
    reference_calls.table, 'reference_gt',
    stemcell_hotspot_tb, dosage_sensitive_gene_tb, cancer_gene_tb,
    gr_info, report_config#, caption = 'reference genotype CNV calls'
)
cat('\n\n')
```

```{r reference_gt_calls.plots, eval = !is.na(ref_id) & include.section('reference_gt_calls.plots'), results='asis'}
make_CNV_plot_section(reference_calls.table, 'reference_gt')
cat('\n\n')
```

```{r regions_of_interest, eval=include_roi_plots, results='asis'}
cat('## Regions of Interest {.tabset .tabset-pills}\n\n')
cat(str_glue(
    'This section includes all regions of interest defined for {sample_id}. ',
    'It uses the same layout as the other CNV call sections.\n\n\n\n'
))

roundup_oom <- function(x) {10^(ceiling(log10(x)))}

source(file.path(config$snakedir, 'scripts/R/CNV_preprocess_functions.R'))
source(file.path(config$snakedir, 'scripts/R/CNV_annotation_functions.R'))

fake_call_tb <- roi_gr %>%
    add_snp_probe_counts(
        raw_LRR_BAF %>% as_granges(seqnames = Chr, start = Position, width = 1)
    ) %>% 
    mutate(
        i = 1:plyranges::n(),
        ID = paste0('ROI_', i),
        FILTER = NA_character_
    ) %>%
    select(-list_name, -mapping, -call_type, -description_doi, -check_score) %>%
    annotate_gaps(
        fix_rel_filepath(config$array_definition[[array]]$array_gaps_file, config),
        config$settings$CNV_processing$call_processing$min.perc.gap_area, 
        config$settings$CNV_processing$call_processing$gap_area.uniq_probes.rel,
        target_chrom_style
    ) %>%
    annotate_high_density(
        fix_rel_filepath(config$array_definition[[array]]$array_density_file, config),
        config$settings$CNV_processing$call_processing$density.quantile.cutoff,
        target_chrom_style
    ) %>%
    annotate_gene_overlaps(gr_genes) %>%
    as_tibble() %>%
    dplyr::rename(chrom = seqnames, Size = width, ROI_hits = hotspot) %>%
    mutate(
        sample_id = sample_id,
        Check_Score = NA_real_,
        CNV_type = 'ROI',
        CN = NA_integer_,
        CNV_caller = 'ROI',
        Call_label =  ifelse(is.na(description), ID, description),
        caller_merging_coverage = NA_real_,
        precision_estimate = NA_character_,
        precision_estimate_description = NA_character_,
        stemcell_hotspot = NA_character_,
        dosage_sensitive_gene = NA_character_,
        cancer_gene = NA_character_,
        LRR = NA_real_, #FIXME: this could be calculated
    )

# Adjust factor levels
fake_call_tb$chrom <- fct_expand(fake_call_tb$chrom, levels(raw_LRR_BAF$Chr))
fake_call_tb$chrom <- fct_relevel(fake_call_tb$chrom, levels(raw_LRR_BAF$Chr))

# Make sure to include all ROIS, which use the Name/ID as a Call Label (instead of the default labels)
report_config$call.data.and.plots$regions_of_interest$always_include_CNVs <- fake_call_tb$Call_label

CNV_table_output(
    fake_call_tb, 'regions_of_interest', 
    stemcell_hotspot_tb, dosage_sensitive_gene_tb, cancer_gene_tb,
    gr_info, report_config#, caption = 'Regions of Interest'
)
make_CNV_plot_section(fake_call_tb, 'regions_of_interest')

cat('\n\n')
```


# SNV analysis {.tabset}

```{r load.SNV.data}

SNV_table <- params$inputs$snv_analysis %>%
    fix_rel_filepath(config) %>%
    read_excel(sheet = 'SNV_table')

gene_coverage_table <- params$inputs$snv_analysis %>%
    fix_rel_filepath(config) %>%
    read_excel(sheet = 'SNV_hotspot_coverage')

snc_qc_tb <- params$inputs$snv_analysis %>%
    fix_rel_filepath(config) %>%
    read_excel(sheet = 'SNP_QC_details')

```

```{r SNV.table, eval = include.section('SNV.table'), results='asis'}
cat('## Table of de-novo SNVs\n\n')

paste0(
    'This table lists all SNVs detected by the Chip Array which are different from the reference genome and ',
    'are annotated as at least protein changing. Due to their potential impact these are now called "SNVs" rather than ',
    '"SNPs", independent of their actual (unknown) frequency in the population.\n\n',
    'All SNVs are categorised into one of the following categories (shown in the hidden SNV category column):\n\n',
    paste(imap(
        defined_labels$SNV_categories, \(x, name) str_glue(' - {name}: {x}')
        ), collapse = '\n\n'
    ),
    '\n\nThe "SNV label" further categorizes the SNVs into:\n\n',
    paste(imap(
        defined_labels$SNV_labels, \(x, name) str_glue(' - {name}: {x}')
        ), collapse = '\n\n'
    ),
    '\n\nThe following categories are assinged as "critical" or "reportable":\n\n',
    paste(imap(
        c(
            setNames(config$settings$SNV_analysis$critical_SNV, rep('critical', length(config$settings$SNV_analysis$critical_SNV))),
            setNames(config$settings$SNV_analysis$reportable_SNV, rep('reportable', length(config$settings$SNV_analysis$reportable_SNV)))
        ),\(x, name) str_glue(' - {name}: {x}')
        ), collapse = '\n\n'
    ),
    '\n\nThe table allows sorting and filtering the SNVs by various criteria, default is sorting by the `SNV Label`.',
    'Hovering over the column headers gives explanations for each column and the "Column visibility" button can be',
    'used to show (or hide) columns. Each SNV genotype (GT) is shown in vcf format: each allele is represented by a ',
    'single number, separated by a forward slash. A 0 indicates the reference allele, a 1 indicates the ',
    'alternate allele. A dot (.) indicates that the genotype could not be determined.'    
) %>% collapsible_html_text('SNV table explanations')

if (is.na(ref_id)) {
    cat(
        'No reference sample is defined therefore all SNVs are labelled included in the "de-novo" table. ',
        'Howeever, the vast majority of SNVs is likely derived from the originating sample/germline.\n\n'
    )
}

SNV_table_output(
    SNV_table %>% filter(SNV_label != 'reference genotype'), 
    roi_gr, snv_hotspot_tb, config, report_config, defined_labels, params$out_format #, caption
)

ref_snv <- SNV_table %>% filter(SNV_label == 'reference genotype')

if (nrow(ref_snv) > 0) {
    cat('## Table of reference SNVs\n\n')
    
    SNV_table_output(
        ref_snv, 
        roi_gr, snv_hotspot_tb, config, report_config, defined_labels, params$out_format #, caption
    )
}

    
cat('\n\n')

```

```{r SNV.hotspot.coverage, eval = include.section('SNV.hotspot.coverage'), results='asis'}
cat('## SNV hotspot coverage\n\n')

paste(
    'This table lists all genes that have known point mutation hotspots for stem cells, a source for the hotspots,',
    'the selected primary transcript for each gene, as well as the coverage of the genes on cDNA, CDS and protein level',
    '(percent coverage of bases/amino acids, as well as absolute numbers).',
    'The coverage is based on all probes contained on the utilised array.\n\n',
    'The "Hotspots" column, lists the specific annotated protein changes for each gene and whether or not any probe on',
    'the array covers each of the specific mutations.'
) %>% collapsible_html_text('SNV hotspot coverage explanations')

full_join(
    snv_hotspot_tb %>% select(gene_name, description_htmllinks) %>% 
        dplyr::rename(description = description_htmllinks) %>%
        unique(),
    gene_coverage_table,
    by = 'gene_name'
) %>%
    mutate(hotspots = str_replace_all(hotspots, '\\n', '<br>')) %>%
    rename_with(format_column_names) %>%
    simple_table_output(params$out_format, escape = FALSE)

cat('\n\n')
```

```{r SNV.QC.details, eval = include.section('SNV.QC.details'), results='asis'}
cat('## SNV QC details\n\n')
snc_qc_tb %>%
    # Move the description of changes on each chromosome into separate columns,
    # so that are displayed properly afetr transforming the table, Do no reformat those column names
    mutate(
        chrom_col_name = str_remove(chromosome, ':.*'),
        chrom_col_val  = str_remove(chromosome, '.*:')
    ) %>%
    select(-chromosome) %>%
    pivot_wider(names_from = chrom_col_name, values_from = chrom_col_val) %>%
    rename_with(format_column_names, 1:4) %>%
    tr_tibble() %>%
    simple_table_output(params$out_format)
cat('\n\n')
```


# Sample comparison

```{r genome.overview, eval = include.section('genome.overview'), results='asis'}

cat('## Genome Overview {.tabset}\n\n')

paste(
    'The following plots each show a whole chromosome overview of the sample, combining to a whole genome view.',
    'CNV calls (filtered based on the config settings) are shown as colored background bars, with the color indicating the',
    'type of call: green for gains, red for losses, and grey for LOH.',
    'Additionally, if the sample has a reference, SNVs that are labelled critical or protein changing/unreliable',
    'critical are also highlighted in red and orange, respectively.\n\n'
) %>% collapsible_html_text('Genome overview explanations')

if (!is.na(ref_id)) {
    label_SNVs <- SNV_table %>%
        filter(SNV_label != 'reference genotype') %>%
        mutate(
            sample_id = sample_id,
            color = ifelse(SNV_label == 'critical', 'red', 'orange')
        )
} else {
    label_SNVs <- SNV_table %>% mutate(color = 'red', sample_id = sample_id) %>% filter(SNV_label == 'dummy')
}

filtered_calls <- cnv_calls %>%
    filter(Call_label %in% report_config$genome_overview$call_labels_overview)

for (chrom in levels(raw_LRR_BAF$Chr) %>% str_subset('M', T)) {
    cat('### ', chrom, '\n\n')
    
    subchunkify(
        make_chromsome_overview_plot(chrom, sample_headers[1], label_SNVs, raw_LRR_BAF, filtered_calls, gr_info),
        paste0('chromosome_overview.', chrom), 5, 12
    )
    
    if (!is.na(ref_id) & report_config$genome_overview$show_reference) {
        subchunkify(
            make_chromsome_overview_plot(chrom, sample_headers[2], label_SNVs, raw_LRR_BAF, filtered_calls, gr_info),
            paste0('chromosome_overview.ref.', chrom), 5, 12
        )
    }
    
    cat('\n\n')
}

cat('\n\n')

```

```{r snp.dendrogram, results='asis', eval = include.section('SNP.dendrogram')}

cat('## Identity comparison\n\n')

snp.distance.tb <-  read_excel(
    fix_rel_filepath(params$input$snv_analysis, config),
    sheet = 'SNP_GT_distances'
)

if (nrow(snp.distance.tb) == 1) {
    
    cat('No additional samples used for comparison, no dendrogram can be built.\n\n')
    
} else {
    
    paste(
        'Sample identities can be comparsed based on the dendrogram built on the SNP genotypes.',
        'The dendrogram is built using the manhattan distance between samples, counting both alleles',
        'from Probes that are not quality in every included sample.',
        'Accordingly, the distance between two samples is the sum of the absolute differences between the two',
        'alleles at each SNP (also shown in the table below) after QC filters.',
        'Samples that are very close together are likely identical or clonally related.',
        'Sample selection as well as color and shape lables are controlled by the config file.\n\n'
    ) %>% collapsible_html_text('Dendrogram explanations')
    
    #Build annotation table
    color_by <- report_config$SNP_comparison$dendrogram.color.by
    shape_by <- report_config$SNP_comparison$dendrogram.shape.by
    dend.format.df <- sampletable %>%
        filter(Sample_ID %in% snp.distance.tb$sample_distance_to)
    
    if (length(color_by) == 1) {
    
        if (nrow(unique(dend.format.df[, color_by])) > 10) {
            warn_str <- str_glue('More than 10 colors are needed to use "{color_by}" for coloring. Consider using fewer unqiue entries.')
            cat(warn_str, '\n\n')
            warning(warn_str)
        }
    
        dend.format.df <- dend.format.df %>%
            mutate(across(all_of(color_by), ~ factor(., levels = sort(unique(.))))	) %>%
            arrange(!!sym(color_by)) %>%
            mutate(col = viridis_pal(option='H')(length(unique(!!sym(color_by))))[match(!!sym(color_by), unique(!!sym(color_by)))],
                         col = ifelse(is.na(!!sym(color_by)), 'grey90', col)
                         )
    
        col_map <- dend.format.df$col %>% unique()
        names(col_map) <- levels(pull(dend.format.df, !!sym(color_by)))
    
    
    } else {
        warning('No matching column to use for coloring!')
        cat('No matching column to use for coloring!\n')
    
        dend.format.df <- dend.format.df %>%
            mutate(col = 'black', `_dummy_color` = 'bar')
    
        col_map  <- c('bar' = 'black')
        color_by <- '_dummy_color'
    }
    
    if (length(shape_by) == 1) {
    
        dend.format.df <- dend.format.df %>%
            mutate(across(all_of(shape_by), ~ factor(., levels = sort(unique(.), na.last = T))),
                         ) %>%
            arrange(!!sym(shape_by))
    
        if (nrow(na.omit(unique(dend.format.df[, shape_by]))) <= 5) {
            use.shapes <- 15:18
            na.shape <- 1
        } else if (nrow(na.omit(unique(dend.format.df[, shape_by]))) <= 15) {
            use.shapes <- 0:14
            na.shape <- 16
        } else {
            others <- dend.format.df[pull(dend.format.df, !!sym(shape_by)) %>% as.integer() > 15, shape_by] %>% na.omit() %>% unlist()
            warn_str <- str_glue('Only 15 shapes are available, but "{shape_by}" would need {length(unique(pull(dend.format.df, !!sym(shape_by))))}. ',
                                 'Consider using fewer unqiue entries. These values are summarised as "Other": {paste(others, collapse = ", ")}')
            cat(warn_str, '\n\n')
            warning(warn_str)
    
            use.shapes <- 0:14
            na.shape <- 16
    
            dend.format.df <- dend.format.df %>%
                mutate(
                    across(all_of(shape_by), ~ ifelse(as.integer(.) > 15, 'Other', as.character(.))),
                    across(all_of(shape_by), ~ factor(., levels = unique(.)))
                )
        }
    
        dend.format.df <- dend.format.df %>%
            mutate(
                shape = use.shapes[match(!!sym(shape_by), unique(!!sym(shape_by)))],
                shape = ifelse(is.na(!!sym(shape_by)), na.shape, shape)
            )
    
        shape_map <- dend.format.df$shape %>% unique()
        names(shape_map) <- levels(pull(dend.format.df, !!sym(shape_by)))    
    } else {
        warning('No matching column to use for shapes!')
        cat('No matching column to use for shapes!\n\n')
    
        dend.format.df <- dend.format.df %>%
            mutate(shape = 16, `_dummy_shape` = 'foo')
    
        shape_map <- c('foo' = 16)
        shape_by <- '_dummy_shape'
    
    }

    SNP.genotype.distances <- snp.distance.tb %>% select(-sample_distance_to) %>% as.matrix()
    
    # Build dendrogram & sort annotation table by it
    hc <- snp.distance.tb %>%
        column_to_rownames('sample_distance_to') %>%
        as.dist() %>%
        hclust()
    dd <- as.dendrogram(hc) 
    dend.format.df <- dend.format.df[match(labels(dd), dend.format.df$Sample_ID),]
    
    dend <- dd %>%
        set('labels_col', dend.format.df$col) %>%
        set('leaves_col', dend.format.df$col) %>%
        set('leaves_pch', dend.format.df$shape) 
    
    gg1 <- dend %>%
        raise.dendrogram (max(SNP.genotype.distances)/50) %>% 
        set('labels_cex', .8) %>%
        set('leaves_cex', 2) %>%
        set('branches_lwd', .5) %>%
        as.ggdend() %>%
        ggplot(offset_labels = -max(SNP.genotype.distances)/25, theme = theme_classic()) +
        scale_y_continuous(
            name = 'SNP distance (manhattan method)',
            breaks = function(limits) breaks_pretty()(c(max(limits[1], 0), limits[2]), 5) + max(SNP.genotype.distances)/50,
            labels = function(breaks) number(breaks - max(SNP.genotype.distances)/50, big.mark = ','),
            limits = c(-max(SNP.genotype.distances)/2, NA),
            expand = expansion()
        ) +
        scale_x_continuous(limits = c(0, NA), expand = expansion(add = c(0, .5))) +
        geom_line(
            data = tibble(x = c(0, 0), y = max(SNP.genotype.distances) * c(1/50, 51/50)),
            aes(x=x, y=y)
        ) +
        theme(
            legend.position = "bottom",
            axis.text.x = element_blank(),
            axis.title.x = element_blank(),
            axis.line.x = element_blank(),
            axis.ticks.x = element_blank(),
            axis.line.y = element_blank(),
            axis.title.y = element_text(angle = 90, hjust = .75)
        )
    
    # make legend
    gg2 <- dend.format.df %>%
        #arrange(SampleGroup) %>%
        ggplot(aes(x = Sample_ID, y = 1, col = !!sym(color_by), shape = !!sym(shape_by))) +
        geom_point() + 
        scale_color_manual(
            values = col_map,
            guide = ifelse(
                color_by == '_dummy_color',
                list('none'),
                list(guide_legend(direction = 'horizontal',title.position = 'top', ncol = 8, byrow=T))
            )[[1]]
        )  +
        scale_shape_manual(
            values = shape_map,
            guide = ifelse(
                shape_by == '_dummy_shape',
                list('none'),
                list(guide_legend(direction = 'horizontal',title.position = 'top', ncol = 8, byrow=T))
            )[[1]]
        ) +
        theme(legend.box = "vertical", legend.text = element_text(size = 10), legend.title = element_text(size = 12))
    
    gg <- gg1 + as_ggplot(get_legend(gg2)) + plot_layout(ncol = 1, heights = c(5, 1))
    
    subchunkify(gg, 'snp.dendrogram.plot', 8, 10)

    snp.distance.tb %>%
        filter(sample_distance_to == sample_id) %>%
        dplyr::select(-sample_distance_to) %>%
        select(all_of(labels(dd))) %>%
        simple_table_output(params$out_format, paste0('Distances to "', sample_id, '"'))

}

```
