---
title: "`r paste(params$sample_id, str_replace_all(params$report_name, '[\\._-]', ' '))`"
author: "`r paste('StemCNV-check, version:',  params$version)`"
output:
  html_document:
    df_print: paged
    toc: true
    toc_float: true
  pdf_document:
    default
params:
  sample_id: ''
  inputs: NULL
  config: NULL
  report_config: NULL
  config_delta: NULL
  report_name: 'StemCNV-check-report'
  out_format: 'html'
  workdir: ''
  basedir: ''
  version: 'unversioned'
---

<style type="text/css">
.main-container {
  max-width: 100% !important;
  margin: auto;
}
.badge-HI {
  background-color:#ff0000;
  box-sizing: border-box;
}
.badge-HL {
  background-color:#ffa500;
  box-sizing: border-box;
}
.nav-pills {
  # visibility: hidden;
  display: none;
}
[role=tab] {
  height: 40px;
  width: 250px;
}
</style>


```{r setup, include=FALSE}
library(plyranges)
library(tidyverse)
library(dendextend)
library(ggpubr)
library(scales)
library(RIdeogram)
library(DT)
library(yaml)
library(patchwork)
library(knitr)
library(kableExtra)
library(sessioninfo)
library(readxl)
`%!in%` <- Negate(`%in%`)

config <- params$config
report_config <- params$report_config 

# Load helper functions
source(file.path(config$snakedir, 'scripts/R/helper_functions.R'))
source(file.path(config$snakedir, 'scripts/R/R_table_functions.R'))
source(file.path(config$snakedir, 'scripts/R/R_plotting_functions.R'))
source(file.path(config$snakedir, 'scripts/R/vcf_io_functions.R'))

# fix_empty_config_lists <- function(report_config) {
#     # covered in config & cehck AND not needed
#     # if (is.null(report_config$include_sections)) {
#     #     report_config$include_sections <- list() 
#     # }
#     # This actually makes it worse
#     # if (is.null(report_config$sample.info.extra.cols)) {
#     #     report_config$sample.info.extra.cols <- list() 
#     # }
#     # if is.null(report_config$ideogram.exclude_filter
#     
#     # Empty lists and NULL will make issues here, might need a 'dummy' instead?
#     for (plotsection in c('denovo', 'reference_gt', 'regions_of_interest')) {
#         if (is.null(report_config$call.data.and.plots[[plotsection]]$always_include)) {
#             report_config$call.data.and.plots[[plotsection]]$always_include <- list()
#         }
#         if (is.null(report_config$call.data.and.plots[[plotsection]]$exclude_filter)) {
#             report_config$call.data.and.plots[[plotsection]]$exclude_filter <- list()
#         }
#     }
#     report_config
# }
# 
# report_config <- params$report_config  %>%
#     # Empty yaml lists are converted to NULL by snakemake
#     fix_empty_config_lists()

include.section <- function(section) {
    include <- ifelse(
        all(report_config$include_sections == '__all__'), 
        T,
        section %in% report_config$include_sections
    )
    exclude <- section %in% report_config$exclude_sections
    # also exclude QC sections if they are not defined in settings
    include <- include & !(
        str_detect(section, '^QC\\.(?!(summary|GenCall|settings))') & 
        str_remove(section, 'QC\\.') %!in% config$settings$CNV.calling.tools
    )
    include & !exclude
}

# General variables & settings

image_folder <- paste0(params$report_name, '-', params$out_format, '_images/')
workdir <- ifelse(params$workdir == '', getwd(), params$workdir)
knitr::opts_chunk$set(
    echo = FALSE, message=NA, warning=NA, error=FALSE,
    fig.path = file.path(workdir, image_folder)
)
options(knitr.kable.NA = '')


sampletable <- config$sample_table %>%
    fix_rel_filepath(config) %>%
    read_tsv(col_types = 'cccccc', comment = "#")

sample_id <- params$sample_id
sex <- get_sample_info(sample_id, 'sex', sampletable)
ref_id <- get_sample_info(sample_id, 'ref_id', sampletable)
sex.ref <- get_sample_info(sample_id, 'sex.ref', sampletable)

#Sample_Name is an optional column in the sampletable
if ('Sample_Name' %in% colnames(sampletable)) {
    sample_name <- sampletable[sampletable$Sample_ID == sample_id, ]$Sample_Name
    ref_name <- ifelse(is.na(ref_id), NA, sampletable[sampletable$Sample_ID == ref_id, ]$Sample_Name)
} else {
    sample_name <- sample_id
    ref_name <- ifelse(is.na(ref_id), NA, ref_id)
}

if (!is.na(ref_id)) {
    sample_headers <- c(sample_name, paste0('Reference (', ref_name, ')'))
    names(sample_headers) <- c(sample_id, ref_id)
    use_ids <- c(sample_id, ref_id)
} else {
    sample_headers <- sample_name
    names(sample_headers) <- sample_id
    use_ids <- sample_id
}

target_chrom_style <- get_target_chrom_style(config, parse_cnv_vcf(params$input$cnv_vcf))
gr_genes <- load_gtf_data(config, target_chrom_style)
gr_info  <- load_genomeInfo(config, target_chrom_style)

roi_regions <- sampletable[sampletable$Sample_ID == sample_id, ]$Regions_of_Interest %>%
    str_split(';') %>% unlist()
include_roi_plots <-  include.section('regions.of.interest') & ifelse(
    'Regions_of_Interest' %in% colnames(sampletable), 
    !is.na(roi_regions) & roi_regions != '', 
    FALSE
)

high_impact_tb <- config$settings$CNV_processing$gene_overlap$high_impact_list %>%
    str_replace('__inbuilt__', config$snakedir) %>%
    fix_rel_filepath(config) %>%
    read_tsv()
highlight_tb <- config$settings$CNV_processing$gene_overlap$highlight_list %>%
    str_replace('__inbuilt__', config$snakedir) %>%
    fix_rel_filepath(config) %>%
    read_tsv()
## general functions

tr_tibble <- function(tb) {
    tr <- t(tb )
    colnames(tr) <- sample_headers[tr[1,]]
    tb <- as_tibble(tr, rownames = ' ')
    tb[2:nrow(tb),]
}

## Rmd related functions

CNV_ID_str <- function(ID, i, region_name = NULL, ...) {
    #ROIs might have user defined names
    if(!is.null(region_name)) id_str <- paste0('roi-', region_name)
    # All other calls
    else id_str <- str_glue('nr{i}-{ID}')
    
    str_to_lower(id_str) %>%
        str_replace_all('[:]', '')
}


format_size <- function(size) {
    oom_f <- ifelse(size >= 1e6, 1e6, 1e3)
    num <- round(size / oom_f)
    paste0(num, ifelse(oom_f == 1e3, 'kb', 'Mb'))
}

make_CNV_plot_section <- function(call.table, plotsection = 'denovo') {
    if (nrow(call.table) == 0) {
        return(str_glue('No {plotsection} CNV calls found.'))
    }
    always_include <- report_config$call.data.and.plots[[plotsection]]$always_include
    for (i in call.table$i) {
        
        row <- call.table[i,]
        res <- make_LRR_BAF_plots(
            row, raw_LRR_BAF, cnv_calls, gr_genes, gr_info,
            report_config$call.data.and.plots[[plotsection]]$plot.region.minsize,
            report_config$call.data.and.plots[[plotsection]]$plot.flanking.region.relative
        )
        if(i <= report_config$call.data.and.plots[[plotsection]]$min_number_plots | row$Call_label %in% always_include) {
            cat('### ', pmap_chr(row, CNV_ID_str), '\n\n')
            if (typeof(res$gg) == 'character') {
                cat(res$gg, '\n\n')
                next()
            }
            if (report_config$call.data.and.plots[[plotsection]]$include.plot) {
                subchunkify(res$gg, str_glue('CNV_call.{plotsection}.nr{i}.plot'), 5, 10)
                cat('\n\n')
            }
            message(str_glue('N hotspots: {length(res$hotspots)}'))
            if (report_config$call.data.and.plots[[plotsection]]$include.hotspot.table & length(res$hotspots)>0) {
                res$hotspots %>%
                    hotspot_table_output(plotsection, high_impact_tb, highlight_tb, report_config, params$out_format) %>%
                    subchunkify(str_glue('CNV_call.{plotsection}.nr{i}.hotspots'))
                cat('\n\n')
            }
            if (report_config$call.data.and.plots[[plotsection]]$include.gene.table.details != 'None') {
                gene_table_output(res$genes, plotsection, high_impact_tb, highlight_tb, report_config) %>%
                    subchunkify(str_glue('CNV_call.{plotsection}.nr{i}.table.genes'))
                cat('\n\n')
            }
            cat('\n\n')
        } else if (typeof(res$gg) != 'character') {
            ggsave(
                file.path(image_folder, str_glue('CNV_call.{plotsection}.nr{i}.plot-1.png')),
                res$gg, 
                width = 20, height = 10, dpi=200
            )
        }
    }
}

# Make a separate subchunnk, this is often needed when mixing loops, tabsets & html elements
subchunkify <- function(g, id, fig_height=7, fig_width=7, options = '') {
  g_deparsed <- paste0(deparse(function() {g}), collapse = '')

  sub_chunk <- paste0("```{r ", id, ", fig.height=", fig_height, ", fig.width=", fig_width,
                      ifelse(options == '', options, paste(',', options)), ", echo=FALSE}\n",
  "(",  g_deparsed, ")()",
  "\n```")

  cat(knitr::knit(text = knitr::knit_expand(text = sub_chunk), quiet = TRUE))
}

```


# Sample Overview

```{r sample.information, eval = include.section('sample.information'), results = 'asis'}

cat('## Sample Information\n\n')

info.tb <- sampletable %>%
  filter(Sample_ID %in% c(sample_id, ref_id)) %>%
  dplyr::select(Sample_ID, Sex, Reference_Sample, any_of(report_config$sample.info.extra.cols)) %>%
  mutate(`Analysis run date` = Sys.Date()) %>%
  tr_tibble() %>%
  dplyr::select(-starts_with('Reference ('))

simple_table_output(info.tb, params$out_format)

cat('\n\n')

```

## QC metrics {.tabset}

```{r qc.summary, eval = include.section('QC.summary'), results='asis'}
cat('### Summary\n\n')
read_excel(
    fix_rel_filepath(params$input$summary_xlsx, config),
    sheet = 'summary_stats'
) %>%
    summary_table(sample_headers, config)
cat('\n\n')
```

```{r qc.gencall, eval = include.section('QC.GenCall'), results='asis'}
cat('### GenCall\n\n')
read_excel(
    fix_rel_filepath(params$input$summary_xlsx, config),
    sheet = 'gencall_stats'
) %>%
    simple_table_output(params$out_format)
cat('\n\n')
```

```{r qc.penncnv, eval = include.section('QC.PennCNV'), results='asis'}
cat('### PennCNV\n\n')
read_excel(
    fix_rel_filepath(params$input$summary_xlsx, config),
    sheet = 'PennCNV_QC_info'
) %>%
    simple_table_output(params$out_format)

read_excel(
    fix_rel_filepath(params$input$summary_xlsx, config),
    sheet = 'PennCNV_stats'
) %>%
    simple_table_output(params$out_format)
cat('\n\n')
```

```{r qc.CBS, eval = include.section('QC.CBS'), results='asis'}
cat('### CBS\n\n')
read_excel(
    fix_rel_filepath(params$input$summary_xlsx, config),
    sheet = 'CBS_stats'
) %>%
    simple_table_output(params$out_format)
cat('\n\n')
```

```{r config, eval = include.section('QC.settings'), results='asis'}
cat('### Config\n\n')
config_delta <- params$config_delta %>%
    lapply(\(sublist) {
        names(sublist) <- names(sublist) %>%
            str_replace("root\\['", "config: ") %>%
            str_replace_all("'\\]\\['", " : ") %>%
            str_remove("'\\]") %>%
            str_replace('report_settings', paste('reports :', params$report_name, ': '))
        sublist
     })

cat('Changes from default config:\n\n')
cat(knitr::knit(text = knitr::knit_expand(text = '```yaml\n`r as.yaml(config_delta)`\n```'), quiet = TRUE))
cat('\n\n')

cat('Complete config used by StemCNV-check:\n\n')
cat(knitr::knit(text = knitr::knit_expand(text = '```yaml\n`r as.yaml(config)`\n```'), quiet = TRUE))
cat('\n\n')
# session_info()
# cat('\n\n')
cat('### R session info\n\n')
```

```{r sessioninfo, eval = include.section('QC.settings')}
session_info()
```


# CNV calling

```{r CNV.call.plot.data}

#TODO:
# - hardcoded chrom style

raw_LRR_BAF <- c(params$input$snp_vcf, params$input$ref_snp_vcf) %>%
    sapply(fix_rel_filepath, config=config) %>%
    lapply(parse_snp_vcf, apply_filter = FALSE) %>%
    bind_ranges() %>%
    fix_CHROM_format(target_chrom_style) %>%
    as_tibble() %>%
    #FIXME / future: adapt this
    dplyr::rename(Chr = seqnames, Position = start, Name = ID,
                  `B Allele Freq` = BAF, `Log R Ratio` = LRR) %>%
    mutate(filter.passed = FILTER == 'PASS') %>%
    dplyr::select(sample_id, Name, Chr, Position, filter.passed, `Log R Ratio`, `B Allele Freq`) 

cnv_calls <- c(params$input$cnv_vcf, params$input$ref_cnv_vcf) %>%
    sapply(fix_rel_filepath, config=config) %>%
    lapply(parse_cnv_vcf) %>%
    bind_ranges() %>%
    fix_CHROM_format(target_chrom_style) %>%
    as_tibble() %>%
    mutate(
        CNV_type = factor(CNV_type),
        reference_overlap = factor(!is.na(reference_coverage)),
        Call_label = ifelse(is.na(Call_label), '', Call_label),
        Call_label = factor(Call_label, levels = c('Reportable', 'Critical', 'Reference genotype', '')),
        FILTER = ifelse(is.na(FILTER), '', FILTER),
        probe_coverage_gap = str_detect(FILTER, 'probe_gap'),
        high_probe_density = str_detect(FILTER, 'high_probe_dens'),
        caller_merging_coverage = sapply(initial_call_details, \(x){
            if (is.na(x)) return(NA)
            str_split(x, '\\|') %>%
                unlist() %>%
                str_remove('[0-9]+-[0-9]+_CN[0-9]_cov') %>%
                str_replace('_[^_]+$', '%') %>%
                str_replace('_', ':') %>%
                paste(collapse = '; ')
        }),
    ) 

#factor levels need to include all chromosomes for comparison
levels(cnv_calls$seqnames) <- levels(raw_LRR_BAF$Chr)


denovo_filter_regex <- ifelse(
    is.null(report_config$call.data.and.plots$denovo$exclude_filter),
    'dummy',
    report_config$call.data.and.plots$denovo$exclude_filter %>% paste(collapse = '|')
)
denovo_calls.table <- cnv_calls %>% 
    dplyr::rename(chrom = seqnames, Size = width) %>%
    arrange(desc(Check_Score), desc(Size), chrom, start) %>%
    filter(sample_id == !!sample_id & Call_label != 'Reference genotype') %>%
    filter(!str_detect(FILTER, denovo_filter_regex)) %>%
    mutate(i = row_number())

refgt_filter_regex <- ifelse(
    is.null(report_config$call.data.and.plots$reference_gt$exclude_filter),
    'dummy',
    report_config$call.data.and.plots$reference_gt$exclude_filter %>% paste(collapse = '|')
)
reference_calls.table <- cnv_calls %>% 
    dplyr::rename(chrom = seqnames, Size = width) %>%
    arrange(desc(Check_Score), desc(Size), chrom, start) %>%
    filter(sample_id == !!sample_id & Call_label == 'Reference genotype') %>%
    filter(!str_detect(FILTER, refgt_filter_regex)) %>%
    mutate(i = row_number())



```

```{r denovo_calls.header, eval = include.section('denovo_calls.table') | include.section('denovo_calls.plots'), results='asis'}
cat('## de-novo CNV calls {.tabset .tabset-pills}\n\n')
cat('This section describes all de-novo CNV calls, meaning calls without a match in the reference sample.\n\n')
```

```{r denovo_calls.table, results='asis', eval = include.section('denovo_calls.table')}
CNV_table_output(denovo_calls.table, 'denovo', high_impact_tb, highlight_tb, gr_info, report_config, caption = 'de-novo CNV calls')
cat('\n\n')
```

```{r denovo_calls.plots, eval = include.section('denovo_calls.plots'), results='asis'}
make_CNV_plot_section(denovo_calls.table, 'denovo')
cat('\n\n')
```
```{r reference_gt_calls.header, eval = !is.na(ref_id) & (include.section('reference_gt_calls.table') | include.section('reference_gt_calls.plots')), results='asis'}
cat('## Reference genotype CNV calls {.tabset .tabset-pills}\n\n')
cat('This section describes all CNV calls that match the reference sample.\n\n')
```

```{r reference_gt_calls.table, results='asis', eval = !is.na(ref_id) & include.section('reference_gt_calls.table')}
CNV_table_output(reference_calls.table, 'reference_gt', high_impact_tb, highlight_tb, gr_info, report_config, caption = 'reference genotype CNV calls')
cat('\n\n')
```

```{r reference_gt_calls.plots, eval = !is.na(ref_id) & include.section('reference_gt_calls.plots'), results='asis'}
make_CNV_plot_section(reference_calls.table, 'reference_gt')
cat('\n\n')
```

```{r regions_of_interest, eval=include_roi_plots, results='asis'}
cat('## Regions of Interest {.tabset}\n\n')
cat(str_glue('This section includes all regions of interest defined for {sample_name}. \n\n\n\n'))

regions <- sampletable[sampletable$Sample_ID == sample_id, ]$Regions_of_Interest %>%
    str_split(';') %>% unlist()

roundup_oom <- function(x) {10^(ceiling(log10(x)))}

# Using all here would swallow problems; the NA check should only be needed for empty entries
if (any(!is.na(regions)) & length(regions)>0) {

    fake_call_tb <- tibble(
        sample_id = sample_id,
        chrom = str_remove(regions, '^[^|]+\\|') %>%
                str_remove(':.*') %>%
                str_remove(regex('chr', ignore_case = T)),
        Start = str_extract(regions, '(?<=:)[0-9]+') %>% as.numeric(),
        End = str_extract(regions, '(?<=-)[0-9]+$') %>% as.numeric(),
        Size = End - Start,
        region_name = str_extract(regions, '^[^|]+\\|') %>% str_remove('\\|'),
        Call_label = 'ROI'
    ) %>%
        mutate(
            chrom = mapSeqlevels(chrom, seqlevelsStyle(cnv_calls$chrom)),
            i = row_number(),
            region_name = ifelse(is.na(region_name), paste('ROI_Region', i), region_name),
            Check_Score = 0,
            ID = region_name,
        )

  # Rmd doesn"t auto-print output inside {}
    print(CNV_table_output(fake_call_tb, 'regions_of_interest', high_impact_tb, highlight_tb, gr_info, report_config, caption = 'Regions of Interest'))

    make_CNV_plot_section(fake_call_tb, 'regions_of_interest')

} else {
    cat('No regions of interest defined.')
}

cat('\n\n')
```

# Sample comparison  {.tabset}

```{r snp.dendrogram, results='asis', eval = include.section('SNP.dendrogram')}

cat('## SNP dendrogram\n\n')

snp.distance.tb <-  read_excel(
    fix_rel_filepath(params$input$summary_xlsx, config),
    sheet = 'SNP_GT_distances'
)

#Build annotation table
color_by <- report_config$SNP_comparison$dendrogram.color.by #%>% check_cols_exits()
shape_by <- report_config$SNP_comparison$dendrogram.shape.by #%>% check_cols_exits()
dend.format.df <- sampletable %>%
    filter(Sample_ID %in% snp.distance.tb$sample_distance_to)

if (length(color_by) == 1) {

    if (nrow(unique(dend.format.df[, color_by])) > 10) {
        warn_str <- str_glue('More than 10 colors are needed to use "{color_by}" for coloring. Consider using fewer unqiue entries.')
        cat(warn_str, '\n\n')
        warning(warn_str)
    }

    dend.format.df <- dend.format.df %>%
        mutate(across(all_of(color_by), ~ factor(., levels = sort(unique(.))))	) %>%
        arrange(!!sym(color_by)) %>%
        mutate(col = viridis_pal(option='H')(length(unique(!!sym(color_by))))[match(!!sym(color_by), unique(!!sym(color_by)))],
                     col = ifelse(is.na(!!sym(color_by)), 'grey90', col)
                     )

    col_map <- dend.format.df$col %>% unique()
    names(col_map) <- levels(pull(dend.format.df, !!sym(color_by)))


} else {
    warning('No matching column to use for coloring!')
    cat('No matching column to use for coloring!\n')

    dend.format.df <- dend.format.df %>%
        mutate(col = 'black', `_dummy_color` = 'bar')

    col_map  <- c('bar' = 'black')
    color_by <- '_dummy_color'
}

if (length(shape_by) == 1) {

    dend.format.df <- dend.format.df %>%
        mutate(across(all_of(shape_by), ~ factor(., levels = sort(unique(.), na.last = T))),
                     ) %>%
        arrange(!!sym(shape_by))

    if (nrow(na.omit(unique(dend.format.df[, shape_by]))) <= 5) {
        use.shapes <- 15:18
        na.shape <- 1
    } else if (nrow(na.omit(unique(dend.format.df[, shape_by]))) <= 15) {
        use.shapes <- 0:14
        na.shape <- 16
    } else {
        others <- dend.format.df[pull(dend.format.df, !!sym(shape_by)) %>% as.integer() > 15, shape_by] %>% na.omit() %>% unlist()
        warn_str <- str_glue('Only 15 shapes are available, but "{shape_by}" would need {length(unique(pull(dend.format.df, !!sym(shape_by))))}. ',
                             'Consider using fewer unqiue entries. These values are summarised as "Other": {paste(others, collapse = ", ")}')
        cat(warn_str, '\n\n')
        warning(warn_str)

        use.shapes <- 0:14
        na.shape <- 16

        dend.format.df <- dend.format.df %>%
            mutate(
                across(all_of(shape_by), ~ ifelse(as.integer(.) > 15, 'Other', as.character(.))),
                across(all_of(shape_by), ~ factor(., levels = unique(.)))
            )
    }

    dend.format.df <- dend.format.df %>%
        mutate(
            shape = use.shapes[match(!!sym(shape_by), unique(!!sym(shape_by)))],
            shape = ifelse(is.na(!!sym(shape_by)), na.shape, shape)
        )

    shape_map <- dend.format.df$shape %>% unique()
    names(shape_map) <- levels(pull(dend.format.df, !!sym(shape_by)))


} else {
    warning('No matching column to use for shapes!')
    cat('No matching column to use for shapes!\n\n')

    dend.format.df <- dend.format.df %>%
        mutate(shape = 16, `_dummy_shape` = 'foo')

    shape_map <- c('foo' = 16)
    shape_by <- '_dummy_shape'

}

SNP.genotype.distances <- snp.distance.tb %>% select(-sample_distance_to) %>% as.matrix()

# Build dendrogram & sort annotation table by it
hc <- snp.distance.tb %>%
    column_to_rownames('sample_distance_to') %>%
    as.dist() %>%
    hclust()
dd <- as.dendrogram(hc) 
dend.format.df <- dend.format.df[match(labels(dd), dend.format.df$Sample_ID),]

dend <- dd %>%
    set('labels_col', dend.format.df$col) %>%
    set('leaves_col', dend.format.df$col) %>%
    set('leaves_pch', dend.format.df$shape) 

gg1 <- dend %>%
    raise.dendrogram (max(SNP.genotype.distances)/50) %>% 
    set('labels_cex', .8) %>%
    set('leaves_cex', 2) %>%
    set('branches_lwd', .5) %>%
    as.ggdend() %>%
    ggplot(offset_labels = -max(SNP.genotype.distances)/25, theme = theme_classic()) +
    scale_y_continuous(
        name = 'SNP distance (manhattan method)',
        breaks = function(limits) breaks_pretty()(c(max(limits[1], 0), limits[2]), 5) + max(SNP.genotype.distances)/50,
        labels = function(breaks) number(breaks - max(SNP.genotype.distances)/50, big.mark = ','),
        limits = c(-max(SNP.genotype.distances)/2, NA),
        expand = expansion()
    ) +
    scale_x_continuous(limits = c(0, NA), expand = expansion(add = c(0, .5))) +
    geom_line(
        data = tibble(x = c(0, 0), y = max(SNP.genotype.distances) * c(1/50, 51/50)),
        aes(x=x, y=y)
    ) +
    theme(
        legend.position = "bottom",
        axis.text.x = element_blank(),
        axis.title.x = element_blank(),
        axis.line.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.line.y = element_blank(),
        axis.title.y = element_text(angle = 90, hjust = .75)
    )

# make legend
gg2 <- dend.format.df %>%
    #arrange(SampleGroup) %>%
    ggplot(aes(x = Sample_ID, y = 1, col = !!sym(color_by), shape = !!sym(shape_by))) +
    geom_point() + 
    scale_color_manual(
        values = col_map,
        guide = ifelse(
            color_by == '_dummy_color',
            list('none'),
            list(guide_legend(direction = 'horizontal',title.position = 'top', ncol = 8, byrow=T))
        )[[1]]
    )  +
    scale_shape_manual(
        values = shape_map,
        guide = ifelse(
            shape_by == '_dummy_shape',
            list('none'),
            list(guide_legend(direction = 'horizontal',title.position = 'top', ncol = 8, byrow=T))
        )[[1]]
    ) +
    theme(legend.box = "vertical", legend.text = element_text(size = 10), legend.title = element_text(size = 12))

gg <- gg1 + as_ggplot(get_legend(gg2)) + plot_layout(ncol = 1, heights = c(5, 1))

subchunkify(gg, 'snp.dendrogram.plot', 8, 10)

snp.distance.tb %>%
    filter(sample_distance_to == sample_id) %>%
    dplyr::select(-sample_distance_to) %>%
    select(all_of(labels(dd))) %>%
    simple_table_output(params$out_format, paste0('Distances to "', sample_id, '"'))

```

```{r virtual.karyotype, eval = include.section('virtual.karyotype'), results='asis'}

cat('## Virtual Karyotype\n\n')

if (!is.na(ref_id)) {
    cat(str_glue('The first plot shows only this sample ({sample_name}), the second a side by side comparison  with the reference sample (left: {sample_name}; right: {ref_name}).\n\n'))
}

# Use prepared static data table (genome build specific, UCSC derived) instead of inbuilt RIdeogram
# data(human_karyotype, package="RIdeogram")
genome_data <- gr_info %>% 
    filter(centromer) %>%
    group_by(size) %>%
    reduce_ranges() %>%
    as.data.frame() %>%
    dplyr::rename(Chr = seqnames, CE_start = start, CE_end = end, End = size) %>%
    mutate(Start = 0, Chr = str_remove(Chr, 'chr')) %>%
    select(Chr, Start, End, CE_start, CE_end)


# ideo_filter_regex <- ifelse(
#     is.null(report_config$ideogram.exclude_filter),
#     'dummy',
#     report_config$ideogram.exclude_filter %>% paste(collapse = '|')
# )
sample.calls <- cnv_calls %>%
    #filter(!str_detect(FILTER, ideo_filter_regex)) %>%
    dplyr::select(seqnames, start, end, CNV_type, sample_id) %>%
    dplyr::rename(Chr = seqnames, Start = start, End = end) %>%
    mutate(Value = ifelse(CNV_type %!in% c('gain', 'loss'), 0, 1),
           Value = ifelse(CNV_type == 'loss', -1, Value),
           Chr = str_remove(Chr, 'chr'))
overlay.tb <- sample.calls %>%
    filter(sample_id == !!sample_id) %>%
    dplyr::select(Chr, Start, End, Value)

#This should not be the case unless there are 0 calls that generated images
if (!dir.exists(file.path(workdir, image_folder))) {
    dir.create(file.path(workdir, image_folder))
}

svg_file1 <- tempfile(fileext = '.svg')
#somehow 'convertSVG' can't deal with absolute paths
png_file1 <- file.path(image_folder, 'virtualKaryotype_ideogram.png')

ideogram(genome_data, overlaid = overlay.tb, colorset1 = c('red', 'grey50', 'green'), output = svg_file1)
convertSVG(svg_file1, file = png_file1)

if (params$out_format == 'html') {
    cat(paste0('<img src="', png_file1, '" width="800" />'))
} else {
    #For reasons tiny-tex/latex executes from the directoy this template is located in (rather than the defined outdir), so relative paths don't work for pdf
    include_graphics(file.path(workdir, png_file1) %>% normalizePath, rel_path = FALSE)
}

if (!is.na(ref_id)) {
    refdata <- sample.calls %>%
        filter(sample_id == ref_id) %>%
        dplyr::select(Chr, Start, End, Value)

    svg_file2 <- tempfile(fileext = '.svg')
    png_file2 <- file.path(image_folder, 'virtualKaryotype_ideogram_vsRef.png')

    ideogram(
        genome_data,
        overlaid = overlay.tb,
        label = refdata,
        label_type = 'heatmap',
        colorset1 = c('red', 'grey50', 'green'),
        colorset2 = c('red', 'grey50', 'green'), 
        output = svg_file2
    )
    convertSVG(svg_file2, file = png_file2)

    if (params$out_format == 'html') {
        cat(paste0('<img src="', png_file2, '" width="800" />'))
    } else {
        include_graphics(file.path(workdir, png_file2) %>% normalizePath, rel_path = FALSE)
    }
}

cat('\n\n')

```
