---
title: "`r paste(params$sample_id, str_replace_all(params$report_name, '[\\._-]', ' '))`"
author: "`r paste('StemCNV-check, version:',  params$version)`"
output:
  html_document:
    df_print: paged
    toc: true
    toc_float: true
  pdf_document:
    default
params:
  base_path: ''
  sample_id: ''
  sampletable: 'sample_table.txt'
  report_name: 'StemCNV-check-report'
  out_format: 'html'
  configfile: 'config.yaml'
  workdir: ''
  version: 'unversioned'
---

<style type="text/css">
.main-container {
  max-width: 100% !important;
  margin: auto;
}
.badge-HI {
  background-color:#ff0000;
  box-sizing: border-box;
}
.badge-HL {
  background-color:#ffa500;
  box-sizing: border-box;
}
[role=tab] {
  height: 40px;
  width: 250px;
}
</style>


```{r setup, include=FALSE}
#setwd(params$workdir)
library(plyranges)
library(tidyverse)
library(dendextend)
library(ggpubr)
library(scales)
library(RIdeogram)
library(DT)
library(yaml)
library(patchwork)
library(knitr)
library(kableExtra)
library(sessioninfo)
`%!in%` <- Negate(`%in%`)

get_script_dir <- function() {
  match <- commandArgs(trailingOnly = FALSE) %>%
    str_subset('--file=') %>%
    str_remove('--file=')
  check <- sys.frames()[[1]]$ofile
  if (length(match) > 0) {
    # Rscript
	return(normalizePath(match) %>% dirname())
  } else if (!is.null(check)) {
    # source'd via R console
    return(normalizePath(check) %>% dirname())
  } else {
    # likely testing in IDE
	return ('StemCNV-check/scripts')
  }
}
source(file.path(get_script_dir(), 'R/R_io_functions.R'))
source(file.path(get_script_dir(), 'R/R_table_functions.R'))
source(file.path(get_script_dir(), 'R/R_plotting_functions.R'))

config_parse <- function(args, report_name) {
  subconfig <- config$reports[[report_name]]
  for (arg in args) {
    subconfig <- subconfig[[arg]]
  }
  subconfig
}

report.setting <- function(...) {
  args <- list(...)
  #Check if retry for '__default__' plot-section is needed
  if (args[[1]] == 'call.data.and.plots') {
    retry_default <- args[[2]] != 'filter-settings'
    retry_args <- args
    retry_args[[2]] <- '__default__'
  } else {
    retry_default <- FALSE
  }
  #Check config settings for this specific report
  out.value <- config_parse(args, params$report_name)
  if (is.null(out.value) & retry_default) out.value <- config_parse(retry_args, params$report_name)
  # Fall back to default report settings
  if (is.null(out.value)) out.value <- config_parse(args, '__default__')
  if (is.null(out.value) & retry_default) out.value <- config_parse(retry_args, '__default__')
  # Setting not found - error
  if (is.null(out.value)) {
    quit(paste0('Tried to find non-existing config settings: reports$',
                params$report_name, "$", paste(args, collapse="$")))
  }
  #Convert to numeric if that makes sense
  if (!any(is.na(suppressWarnings(as.numeric(out.value))))) {
    out.value <- as.numeric(out.value)
  }
  out.value
}


include.section <- function(section) {
  include <- ifelse(all(report.setting('include_sections') == '__all__'), T, section %in% report.setting('include_sections'))
  exclude <- section %in% report.setting('exclude_sections')
  # also exclude QC sections if they are not defined in settings
  include <- include & !(str_detect(section, '^QC\\.(?!(summary|GenCall|settings))') & str_remove(section, 'QC\\.') %!in% cnv_tools)
  include & !exclude
}


# General variables & settings

image_folder <- paste0(params$report_name, '-', params$out_format, '_images/')

basepath <- params$base_path
workdir <- ifelse(params$workdir == '', getwd(), params$workdir)
knitr::opts_chunk$set(echo = FALSE, message=FALSE, warning=FALSE,
                      fig.path = file.path(workdir, image_folder)
)
options(knitr.kable.NA = '')

config <- read_yaml(params$configfile)
datapath <- dirname(workdir)

cnv_tools <- config$settings$CNV.calling.tools

sampletable <- read_tsv(params$sampletable, col_types = 'cccccc', comment = "#")

sample_id <- params$sample_id
sex <- get_sample_info(sample_id, 'sex', sampletable)
ref_id <- get_sample_info(sample_id, 'ref_id', sampletable)
sex.ref <- get_sample_info(sample_id, 'sex.ref', sampletable)

#Sample_Name is an optional column in the sampletable
if ('Sample_Name' %in% colnames(sampletable)) {
  sample_name <- sampletable[sampletable$Sample_ID == sample_id, ]$Sample_Name
  ref_name <- ifelse(is.na(ref_id), NA, sampletable[sampletable$Sample_Name == ref_id, ]$Sample_Name)
} else {
  sample_name <- sample_id
  ref_name <- ifelse(is.na(ref_id), NA, ref_id)
}

if (!is.na(ref_id)) {
  sample_headers <- c(sample_name, paste0('Reference (', ref_name, ')'))
  names(sample_headers) <- c(sample_id, ref_id)
  use_ids <- c(sample_id, ref_id)
} else {
  sample_headers <- sample_name
  names(sample_headers) <- sample_id
  use_ids <- sample_id
}

#Warning tresholds
red_warnings <- report.setting('SampleQCwarnings', 'warning_level2_red')
callrate_warnings <- report.setting('SampleQCwarnings', 'call_rate')
SNP_distance_to_ref_warning <- report.setting('SampleQCwarnings', 'SNP_distance_to_reference')
ratio_CNV_warnings <- report.setting('SampleQCwarnings', 'loss_gain_log2ratio')
total_CNV_warnings <- report.setting('SampleQCwarnings', 'total_calls_CNV')
total_LOH_warnings <- report.setting('SampleQCwarnings', 'total_calls_LOH')
reportable_CNV_warnings <- report.setting('SampleQCwarnings', 'reportable_calls_CNV')
reportable_LOH_warnings <- report.setting('SampleQCwarnings', 'reportable_calls_LOH')
critical_CNV_warnings <- report.setting('SampleQCwarnings', 'critical_calls_CNV')
critical_LOH_warnings <- report.setting('SampleQCwarnings', 'critical_calls_LOH')

valid_name <- config$wildcard_constraints$sample_id
# valid_name <- ifelse(is.null(valid_name), '[0-9]{12}_R[0-9]{2}C[0-9]{2}', valid_name)
if (!all(str_detect(use_ids, valid_name))) {stop('Sample id does not match supplied or default wildcard constraints!')}

SNP.extra.ids <- c()
if (include.section('SNP.dendrogram')) {
  SNP.extra.ids <- get_SNP_clustering_IDs(report.setting('SNP_comparison', 'extra_samples'), sample_id, sampletable)
} 

gr_genes <- load_gtf_data(config)
gr_info  <- load_genomeInfo(config)

regions <- sampletable[sampletable$Sample_ID == sample_id, ]$Regions_of_Interest %>%
    str_split(';') %>% unlist()
include_roi_plots <-  include.section('regions.of.interest') & ifelse('Regions_of_Interest' %in% colnames(sampletable), !is.na(regions) & regions != '', FALSE)

high_impact_tb <- config$settings$CNV_processing$gene_overlap$high_impact_list %>%
        str_replace('__inbuilt__', config$snakedir) %>%
        read_tsv()
highlight_tb <- config$settings$CNV_processing$gene_overlap$highlight_list %>%
        str_replace('__inbuilt__', config$snakedir) %>%
        read_tsv()
## general functions

tr_tibble <- function(tb) {
  tr <- t(tb )
  colnames(tr) <- sample_headers[tr[1,]]
  tb <- as_tibble(tr, rownames = ' ')
  tb[2:nrow(tb),]
}

## Rmd related functions

CNV_ID_str <- function(Call_Label, CNV_type, Chr, Size, i,
                       region_name = NULL,
                       formatting = 'nice', ...) {
  #ROIs might have user defined names
  if(!is.null(region_name) & formatting == 'nice') return(paste0('ROI: ', region_name))
  else if(!is.null(region_name)) return(paste0('ROI-',region_name))
  # All other calls
  if (formatting == 'nice') {
    return(str_glue('Nr-{i} {CNV_type}: {Chr} - {format_size(Size)}'))
  } else {
    return(str_glue('nr-{i}-{CNV_type}-{Chr}---{format_size(Size)}') %>%
             str_to_lower())
  }
}


format_size <- function(size) {
  oom_f <- ifelse(size >= 1e6, 1e6, 1e3)
  num <- round(size / oom_f)
  paste0(num, ifelse(oom_f == 1e3, 'kb', 'Mb'))
}

make_CNV_plot_section <- function(call.table, plotsection = 'denovo') {
  if (nrow(call.table) == 0) {
    return(str_glue('No {plotsection} CNV calls found.'))
  }
  always_include <- report.setting('call.data.and.plots', plotsection, 'always_include')
  for (i in call.table$i) {
    row <- call.table[i,]
    res <- make_LRR_BAF_plots(row,
                              raw_LRR_BAF, cnv_calls, gr_genes, gr_info,
                              report.setting('call.data.and.plots', plotsection, 'plot.region.minsize'),
                              report.setting('call.data.and.plots', plotsection, 'plot.flanking.region.relative')
                              )
    if(i <= report.setting('call.data.and.plots', plotsection, 'min_number_plots') | row$Call_Label %in% always_include) {
      cat('### ', pmap_chr(row, CNV_ID_str, formatting = 'nice'), '\n\n')
      if (typeof(res$gg) == 'character') {
        cat(res$gg, '\n\n')
          next()
      }
      if (report.setting('call.data.and.plots', plotsection, 'include.plot')) {
        subchunkify(res$gg, str_glue('CNV_call.{plotsection}.nr{i}.plot'), 5, 10)
        cat('\n\n')
      }
      if (report.setting('call.data.and.plots', plotsection, 'include.call.table')) {
        subchunkify(simple_table_output(res$calls), str_glue('CNV_call.{plotsection}.nr{i}.table'))
        cat('\n\n')
      }
      if (report.setting('call.data.and.plots', plotsection, 'include.gene.table.details') != 'None') {
        gene.tb <-  res$genes
        subchunkify(gene_table_output(gene.tb, plotsection, high_impact_tb, highlight_tb),
                    str_glue('CNV_call.{plotsection}.nr{i}.table.genes'))
        cat('\n\n')
      }
      cat('\n\n')
    } else if (typeof(res$gg) != 'character') {
        ggsave(file.path(image_folder, str_glue('CNV_call.{plotsection}.nr{i}.plot-1.png')), res$gg, width = 20, height = 10, dpi=200)
    }
  }
}

# Make a separate subchunnk, this is often needed when mixing loops, tabsets & html elements
subchunkify <- function(g, id, fig_height=7, fig_width=7, options = '') {
  g_deparsed <- paste0(deparse(function() {g}), collapse = '')

  sub_chunk <- paste0("```{r ", id, ", fig.height=", fig_height, ", fig.width=", fig_width,
                      ifelse(options == '', options, paste(',', options)), ", echo=FALSE}\n",
  "(",  g_deparsed, ")()",
  "\n```")

  cat(knitr::knit(text = knitr::knit_expand(text = sub_chunk), quiet = TRUE))
}

```

```{r read.data}

# Read Data

## Raw (BAF / LRR), with column for applied filter
plot.probe.filter.setting <- ifelse(report.setting('call.data.and.plots', 'filter-settings') == '__default__',
                               config$settings$`default-filter-set`,
                               report.setting('call.data.and.plots', 'filter-settings'))

raw_LRR_BAF <- file.path(datapath, use_ids, paste0(use_ids, '.processed-data.tsv')) %>%
  lapply(read_raw) %>%
  bind_rows() %>%
  merge(lapply(file.path(datapath, use_ids, paste0(use_ids, '.filtered-data-', plot.probe.filter.setting, '.tsv')), read_raw) %>%
            lapply(function(x) mutate(x, filter.passed = TRUE) %>% dplyr::select(sample_id, Name, Chr, Position, filter.passed)) %>%
            bind_rows(),
        by = c('sample_id', 'Name', 'Chr', 'Position'), all.x = T
  ) %>%
  mutate(filter.passed = ifelse(is.na(filter.passed), F, filter.passed))

## RAW Genotype info
read_snps <- function(fname) {
    read_tsv(fname, col_select = c(1, 2, 11), show_col_types = FALSE) %>%
        rename_with(~ str_remove(., '^.*\\.gencall\\.'), c(3)) %>%
        mutate(GType = ifelse(GType == 'NC', NA, str_count(GType, 'A')),
               sample_id = basename(fname) %>% str_remove('\\.(processed|filtered)-data.*\\.tsv$'))
}

snp.ids <- unique(c(use_ids, SNP.extra.ids))
cluster.probe.filter.setting <- ifelse(report.setting('SNP_comparison', 'filter-settings') == '__default__',
                               config$settings$`default-filter-set`,
                               report.setting('SNP_comparison', 'filter-settings'))
# hclust will fail with <2 samples, so need to disable that section
if (length(snp.ids) < 2) {
  config$reports[[params$report_name]]$exclude_sections <- c(report.setting('exclude_sections'), 'SNP.dendrogram')
}

ignore.filter <- report.setting('SNP_comparison', 'ignore_filter')
snp_file_pattern <- ifelse(ignore.filter,
                           '.processed-data.tsv',
                           paste0('.filtered-data-', cluster.probe.filter.setting, '.tsv'))
# This is always needed (to get snp_dist_to_ref)
SNP.genotype.distances <- file.path(datapath, snp.ids, paste0(snp.ids, snp_file_pattern)) %>%
  lapply(read_snps) %>%
    bind_rows() %>%
    pivot_wider(names_from = sample_id, values_from = GType, values_fill = NA) %>%
    filter(if_all(everything(), ~!is.na(.))) %>%
    dplyr::select(-Index, -Name) %>%
    t() %>%
    dist(method = 'manhattan')


## CNV calls

cnv_calls <- lapply(file.path(datapath, use_ids, paste0(use_ids, '.combined-cnv-calls.tsv')),
                    load_preprocessed_cnvs) %>%
  bind_rows() %>%
  mutate(
    Call_Label = case_when(
            reference_overlap            				              ~ 'Reference genotype',
            `Check-Score` >= report.setting('impact.score.critical')   ~ 'Critical',
            `Check-Score` >= report.setting('impact.score.reportable') ~ 'Reportable',
            TRUE                         				              ~ ''
    )
  )

## GenCall logs
GenCall.stats <-  file.path(datapath, use_ids, paste0(use_ids, '.stats.txt')) %>%
    lapply(read_tsv, show_col_types = FALSE, col_types = cols(computed_gender = col_character(),
                                                              sentrix_barcode = col_character())) %>%
    bind_rows() %>%
    dplyr::select(-(5:7)) %>%
    mutate(gtc = str_remove(gtc, '.gencall.gtc')) %>%
    dplyr::rename(sample_id = gtc)

```


# Sample Overview


```{r sample.information, eval = include.section('sample.information'), results = 'asis'}

cat('## Sample Information\n\n')

info.tb <- sampletable %>%
  filter(Sample_ID %in% c(sample_id, ref_id)) %>%
  dplyr::select(Sample_ID, Sex, Reference_Sample,
                any_of(report.setting('sample.info.extra.cols'))) %>%
  mutate(`Analysis run date` = Sys.Date()) %>%
  tr_tibble() %>%
  dplyr::select(-starts_with('Reference ('))

simple_table_output(info.tb)

cat('\n\n')

```

## QC metrics {.tabset}

```{r qc.metrics}

metrics <- list(
    GenCall.stats %>%
        dplyr::select(sample_id, call_rate, computed_gender) %>%
        mutate(call_rate = round(call_rate, 3)),
    raw_LRR_BAF %>%
        group_by(sample_id) %>%
        summarise(SNPs_post_filter = (100 * sum(filter.passed, na.rm=T) / unique(GenCall.stats$number_snps) ) %>%
                                                                        format(digits = 2, nsmall=2) %>% paste('%')),
    tibble(sample_id = use_ids,
           SNP_distance_to_reference = unlist(ifelse(is.na(ref_id), list('-'), list(c(SNP.genotype.distances[1], '-'))))
    ),
    cnv_calls %>%
        filter(caller_merging_state != 'pre-overlap') %>%
        group_by(sample_id) %>%
        mutate(class = factor(ifelse(CNV_type %!in% c('gain', 'loss'), 'LOH', 'CNV'), levels = c('CNV', 'LOH')),
               loss_gain_log2ratio = log2(sum(CNV_type == 'gain') / sum(CNV_type == 'loss')) %>% round(digits = 2),
               loss_gain_log2ratio = ifelse(is.infinite(loss_gain_log2ratio) | is.nan(loss_gain_log2ratio), 'div/0', loss_gain_log2ratio),
               # highlight_calls = sum(!is.na(highlight_hits)),
               # high_impact_calls = sum(!is.na(high_impact_hits))
        ) %>%
        group_by(sample_id, loss_gain_log2ratio,
                 #highlight_calls, high_impact_calls,
                 class, .drop = F) %>%
        summarise(total_calls = n(),
                  reportable_calls = sum(Call_Label == 'Reportable'),
                  critical_calls = sum(Call_Label == 'Critical'),
                  ) %>%
        pivot_wider(names_from = class, values_from = matches('(total|reportable|critical)_calls'))
)

# Datatable takes colors differently than ggplot, doesn't work somehow
green_color <- ifelse(params$out_format == 'html', 'rgb(146,208,80)', 'green') #rgb(146,208,80) // lightgreen

warn_color <- function(s) ifelse(s %in% red_warnings, 'red', 'orange')
apply_greq_th <- function(data, warning_levels, second = 'orange'){
   second_color <- ifelse(second %in% c('orange', 'red'), second, warn_color(second))
    case_when(
        data >= warning_levels[[2]] ~ second_color,
        data >= warning_levels[[1]] ~ 'yellow',
        TRUE                       ~ green_color,
    )
}

metric.colors <- list(
    mutate(metrics[[1]],
                 call_rate = case_when(
                    call_rate < callrate_warnings[[2]]  ~ warn_color('call_rate'),
                    call_rate < callrate_warnings[[1]]  ~ 'yellow',
                    TRUE ~ green_color),
                 computed_gender = ifelse(tolower(computed_gender) != sex, 'red', green_color)),
    mutate(metrics[[2]],
                 SNPs_post_filter = 'white'),
    mutate(metrics[[3]],
                 SNP_distance_to_reference = ifelse(SNP_distance_to_reference == '-', 'white',
                                              apply_greq_th(as.integer(SNP_distance_to_reference), SNP_distance_to_ref_warning,'SNP_distance_to_ref_warning')
                 )),
    mutate(metrics[[4]],
                 loss_gain_log2ratio = ifelse(loss_gain_log2ratio == 'div/0', 'white',
                    apply_greq_th(abs(as.numeric(loss_gain_log2ratio)), ratio_CNV_warnings, 'loss_gain_log2ratio')),
                 total_calls_CNV = apply_greq_th(total_calls_CNV, total_CNV_warnings, 'total_calls_CNV'),
                 total_calls_LOH = apply_greq_th(total_calls_LOH, total_LOH_warnings, 'total_calls_LOH'),
                 reportable_calls_CNV = apply_greq_th(reportable_calls_CNV, reportable_CNV_warnings, 'reportable_calls_CNV'),
                 reportable_calls_LOH = apply_greq_th(reportable_calls_LOH, reportable_LOH_warnings, 'reportable_calls_LOH'),
                 critical_calls_CNV = apply_greq_th(critical_calls_CNV, critical_CNV_warnings, 'critical_calls_CNV'),
                 critical_calls_LOH = apply_greq_th(critical_calls_LOH, critical_LOH_warnings, 'critical_calls_LOH'),
                )
        )

not_enabled_label <- c('dummy',
                       ifelse(report.setting('impact.score.critical') == 'NA', 'critical', 'dummy'),
                       ifelse(report.setting('impact.score.reportable') == 'NA', 'reportable', 'dummy')
  ) %>% unique %>% paste(collapse = '|')

Combined.metrics <- purrr::reduce(metrics, full_join, by = 'sample_id') %>%
  tr_tibble() %>%
  filter(!str_detect(` `, not_enabled_label)) %>%
  mutate(` ` = format_column_names(` `))
Combined.colors <- purrr::reduce(metric.colors, full_join, by = 'sample_id') %>%
  tr_tibble() %>%
  filter(!str_detect(` `, not_enabled_label))

get_call_stats_tool <- function(use.tool) {
    cnv_calls %>%
        filter(caller_merging_state != 'combined' & use.tool %in% CNV_caller) %>%
        group_by(sample_id) %>%
        mutate(loss_gain_log2ratio = log2(sum(CNV_type == 'gain') / sum(CNV_type == 'loss')) %>% round(digits = 2),
               loss_gain_log2ratio = ifelse(is.infinite(loss_gain_log2ratio) | is.nan(loss_gain_log2ratio), NA, loss_gain_log2ratio),
               highlight_calls = sum(!is.na(highlight_hits)),
               high_impact_calls = sum(!is.na(high_impact_hits))
        ) %>%
        group_by(sample_id, CNV_type, loss_gain_log2ratio) %>%
        summarise(total_calls = n(),
                  reportable_calls = sum(Call_Label == 'Reportable'),
                  critical_calls = sum(Call_Label == 'Critical')
        ) %>%
        pivot_wider(names_from = CNV_type, values_from = matches('(total|reportable|critical)_calls')) %>%
        dplyr::select(-matches(not_enabled_label)) %>%
        tr_tibble()
}

export.tb <- full_join(
        purrr::reduce(metrics, full_join, by = 'sample_id') %>%
            mutate(across(.cols = everything(), ~as.character(.)))  %>%
        pivot_longer(-sample_id, names_to = 'Measure', values_to = 'Value'),
        purrr::reduce(metric.colors, full_join, by = 'sample_id') %>%
            mutate(across(-sample_id, ~str_replace(., 'white', 'NA') %>%
                                    str_replace(green_color, 'OK') %>%
                                    str_replace('yellow', 'unusual') %>%
                                    str_replace('orange', 'warning') %>%
                                    str_replace('red', 'critical'))) %>%
        pivot_longer(-sample_id, names_to = 'Measure', values_to = 'Check'),
        by = c('sample_id', 'Measure')
        ) %>%
        filter(sample_id == !!sample_id) %>%
        rename_with(format_column_names)

write_tsv(export.tb, file.path(workdir, str_glue('{sample_id}.summary-check.tsv')))

```


```{r qc.summary, eval = include.section('QC.summary'), results='asis'}
cat('### Summary\n\n')
summary_table(Combined.metrics, Combined.colors, sample_headers)
cat('\n\n')
```




```{r qc.gencall, eval = include.section('QC.GenCall'), results='asis'}
cat('### GenCall\n\n')
simple_table_output(GenCall.stats %>% tr_tibble())
cat('\n\n')
```

```{r qc.penncnv, eval = include.section('QC.PennCNV'), results='asis'}
cat('### PennCNV\n\n')
# Extract qc summary stats from PennCNV log files
logpath <- ifelse(fs::is_absolute_path(config$log_path),
                 config$log_path,
                 file.path(basepath, config$log_path))
logfiles <- list.files(file.path(logpath, 'PennCNV', sample_id),
                       pattern = '(auto|chrx|chry).error.log', full.names = T)

penncnv.stats <- lapply(logfiles, function(fname) {
    chrs <- basename(fname) %>% str_extract('^[^\\.]+')
    lines <- readLines(fname)
    # Extract from summary line
    tb <- lines %>%
      str_subset('NOTICE: quality summary') %>%
            str_remove('.*: ') %>%
            str_split(' ') %>%
            unlist() %>%
            as_tibble_col(column_name = 'dummy') %>%
            separate(dummy, c('Name', 'value'), sep = '=') %>%
            filter(Name %!in% c('WF', 'GCWF')) %>%
            mutate(Name = str_remove(Name, '[XY]'))
    # Get wave correction
    tb <- bind_rows(tb, lines %>%
            str_subset('Adjusting LRR by GC model') %>%
            str_remove('.*: ') %>%
            str_split(', ') %>%
            unlist() %>%
            as_tibble_col(column_name = 'dummy') %>%
            separate(dummy, c('Name', 'value'), sep = ' changes from ') %>%
            mutate(Name = paste(Name, '(adjusted)'))
    )
    # Get median correction BAF & LRR (auto only)
    if (chrs == 'auto') {
        tb[str_detect(tb$Name, 'median' ),] <- lines %>%
                str_subset('Median-adjusting') %>%
                str_replace('.*(LRR|BAF).*( -?[0-9]\\.[0-9]{4})', '\\1_median;0-adjusted by\\2') %>%
                as_tibble() %>%
                separate(value, c('Name', 'value'), sep = ';')
    }

    mutate(tb, chr = ifelse(chrs == 'auto', 'chr1:22', chrs))
    }) %>%
    bind_rows() %>%
    pivot_wider(names_from = chr, values_from = value) %>%
    dplyr::rename(` ` = Name)

simple_table_output(penncnv.stats)

# Number of CNVs, LOHs called at which thresholds
tbout <- get_call_stats_tool('PennCNV')
simple_table_output(tbout)
cat('\n\n')
```
```{r qc.CBS, eval = include.section('QC.CBS'), results='asis'}
cat('### CBS\n\n')
tbout <- get_call_stats_tool('CBS')
simple_table_output(tbout)
cat('\n\n')
```

```{r sessioninfo, eval = include.section('QC.settings'), results='asis'}
cat('### Pipeline settings\n\n')
cat(knitr::knit(text = knitr::knit_expand(text = '```yaml\n`r xfun::file_string(params$configfile)`\n```'), quiet = TRUE))
cat('\n\n')
session_info()
cat('\n\n')
```


# CNV calling

```{r CNV.call.functions}

cnv_calls <- cnv_calls %>%
  ungroup() %>%
  dplyr::rename(Chr = seqnames, Size = length, Start = start, End = end) %>%
  mutate(CNV_type = factor(CNV_type),
         reference_overlap = factor(reference_overlap),
  ) %>%
  arrange(desc(`Check-Score`), desc(Size), Chr, Start)

denovo_calls.table <- cnv_calls %>%
  filter(caller_merging_state != 'pre-overlap'  &
           sample_id == !!sample_id &
           Call_Label != 'Reference genotype'
  ) %>%
  mutate(i = row_number())

reference_calls.table <- cnv_calls %>%
  filter(caller_merging_state != 'pre-overlap' &
           sample_id == !!sample_id &
           Call_Label == 'Reference genotype'
  ) %>%
  mutate(i = row_number())

```

```{r denovo_calls.header, eval = include.section('denovo_calls.table') | include.section('denovo_calls.plots'), results='asis'}
cat('## de-novo CNV calls {.tabset}\n\n')
cat('This section describes all de-novo CNV calls, meaning calls without a match in the reference sample.\n\n')
```

```{r denovo_calls.table, results='asis', eval = include.section('denovo_calls.table')}
CNV_table_output(denovo_calls.table, 'denovo', high_impact_tb, highlight_tb, caption = 'de-novo CNV calls')
cat('\n\n')
```

```{r denovo_calls.plots, eval = include.section('denovo_calls.plots'), results='asis'}
make_CNV_plot_section(denovo_calls.table, 'denovo')
cat('\n\n')
```
```{r reference_gt_calls.header, eval = !is.na(ref_id) & (include.section('reference_gt_calls.table') | include.section('reference_gt_calls.plots')), results='asis'}
cat('## Reference genotype CNV calls {.tabset}\n\n')
cat('This section describes all CNV calls that match the reference sample.\n\n')
```

```{r reference_gt_calls.table, results='asis', eval = !is.na(ref_id) & include.section('reference_gt_calls.table')}
CNV_table_output(reference_calls.table, 'reference_gt', high_impact_tb, highlight_tb, caption = 'reference genotype CNV calls')
cat('\n\n')
```

```{r reference_gt_calls.plots, eval = !is.na(ref_id) & include.section('reference_gt_calls.plots'), results='asis'}
make_CNV_plot_section(reference_calls.table, 'reference_gt')
cat('\n\n')
```

```{r regions_of_interest, eval=include_roi_plots, results='asis'}
cat('## Regions of Interest {.tabset}\n\n')
cat(str_glue('This section includes all regions of interest defined for {sample_name}. \n\n\n\n'))

regions <- sampletable[sampletable$Sample_ID == sample_id, ]$Regions_of_Interest %>%
    str_split(';') %>% unlist()

use_chr_str <- all(str_detect(cnv_calls$Chr, 'chr'))
roundup_oom <- function(x) {10^(ceiling(log10(x)))}

# Using all here would swallow problems; the NA check should only be needed for empty entries
if (any(!is.na(regions)) & length(regions)>0) {

    fake_call_tb <- tibble(
            sample_id = sample_id,
            Chr = str_remove(regions, '^[^|]+\\|') %>%
                    str_remove(':.*') %>%
                    str_remove(regex('chr', ignore_case = T)),
            Start = str_extract(regions, '(?<=:)[0-9]+') %>% as.numeric(),
            End = str_extract(regions, '(?<=-)[0-9]+$') %>% as.numeric(),
            Size = End - Start,
            region_name = str_extract(regions, '^[^|]+\\|') %>% str_remove('\\|'),
            Call_Label = 'ROI'
            ) %>%
      mutate(Chr = ifelse(use_chr_str,  paste0('chr', Chr), Chr),
             i = row_number(),
             region_name = ifelse(is.na(region_name), paste('Region', i), region_name),
             `Check-Score` = 0,
             ID = region_name,
             ) %>%
      bind_rows(expected_final_tb)

  # Rmd doesn"t auto-print output inside {}
  print(CNV_table_output(fake_call_tb, 'regions_of_interest', high_impact_tb, highlight_tb, caption = 'Regions of Interest'))

  make_CNV_plot_section(fake_call_tb, 'regions_of_interest')

} else {
  cat('No regions of interest defined.')
}

cat('\n\n')
```

# Sample comparison  {.tabset}

```{r snp.dendrogram, results='asis', eval = include.section('SNP.dendrogram')}

cat('## SNP dendrogram\n\n')

#Build annotation table
color_by <- report.setting('SNP_comparison', 'dendrogram.color.by') #%>% check_cols_exits()
shape_by <- report.setting('SNP_comparison', 'dendrogram.shape.by') #%>% check_cols_exits()
dend.format.df <- sampletable %>%
    filter(Sample_ID %in% snp.ids)

if (length(color_by) == 1) {

    if (nrow(unique(dend.format.df[, color_by])) > 10) {
        warn_str <- str_glue('More than 10 colors are needed to use "{color_by}" for coloring. Consider using fewer unqiue entries.')
        cat(warn_str, '\n\n')
        warning(warn_str)
    }

    dend.format.df <- dend.format.df %>%
        mutate(across(all_of(color_by), ~ factor(., levels = sort(unique(.))))	) %>%
        arrange(!!sym(color_by)) %>%
        mutate(col = viridis_pal(option='H')(length(unique(!!sym(color_by))))[match(!!sym(color_by), unique(!!sym(color_by)))],
                     col = ifelse(is.na(!!sym(color_by)), 'grey90', col)
                     )

    col_map <- dend.format.df$col %>% unique()
    names(col_map) <- levels(pull(dend.format.df, !!sym(color_by)))


} else {
    warning('No matching column to use for coloring!')
    cat('No matching column to use for coloring!\n')

    dend.format.df <- dend.format.df %>%
        mutate(col = 'black', `_dummy_color` = 'bar')

    col_map  <- c('bar' = 'black')
    color_by <- '_dummy_color'
}

if (length(shape_by) == 1) {

    dend.format.df <- dend.format.df %>%
        mutate(across(all_of(shape_by), ~ factor(., levels = sort(unique(.), na.last = T))),
                     ) %>%
        arrange(!!sym(shape_by))

    if (nrow(na.omit(unique(dend.format.df[, shape_by]))) <= 5) {
        use.shapes <- 15:18
        na.shape <- 1
    } else if (nrow(na.omit(unique(dend.format.df[, shape_by]))) <= 15) {
        use.shapes <- 0:14
        na.shape <- 16
    } else {
        others <- dend.format.df[pull(dend.format.df, !!sym(shape_by)) %>% as.integer() > 15, shape_by] %>% na.omit() %>% unlist()
        warn_str <- str_glue('Only 15 shapes are available, but "{shape_by}" would need {length(unique(pull(dend.format.df, !!sym(shape_by))))}. ',
							 'Consider using fewer unqiue entries. These values are summarised as "Other": {paste(others, collapse = ", ")}')
        cat(warn_str, '\n\n')
        warning(warn_str)

        use.shapes <- 0:14
        na.shape <- 16

        dend.format.df <- dend.format.df %>%
            mutate(across(all_of(shape_by), ~ ifelse(as.integer(.) > 15, 'Other', as.character(.))),
                         across(all_of(shape_by), ~ factor(., levels = unique(.))) )
    }

    dend.format.df <- dend.format.df %>%
        mutate(shape = use.shapes[match(!!sym(shape_by), unique(!!sym(shape_by)))],
                     shape = ifelse(is.na(!!sym(shape_by)), na.shape, shape)
                     )

    shape_map <- dend.format.df$shape %>% unique()
    names(shape_map) <- levels(pull(dend.format.df, !!sym(shape_by)))


} else {
    warning('No matching column to use for shapes!')
    cat('No matching column to use for shapes!\n\n')

    dend.format.df <- dend.format.df %>%
        mutate(shape = 16, `_dummy_shape` = 'foo')

    shape_map <- c('foo' = 16)
    shape_by <- '_dummy_shape'

}

# Build dendrogram & sort annotation table by it
hc <- hclust(SNP.genotype.distances)
dd <- as.dendrogram(hc) 
dend.format.df <- dend.format.df[match(labels(dd), dend.format.df$Sample_ID),]

dend <- dd %>%
  set('labels_col', dend.format.df$col) %>%
  set('leaves_col', dend.format.df$col) %>%
  set('leaves_pch', dend.format.df$shape) 

gg1 <- dend %>%
  raise.dendrogram (max(SNP.genotype.distances)/50) %>% 
  set('labels_cex', .8) %>%
  set('leaves_cex', 2) %>%
  set('branches_lwd', .5) %>%
  as.ggdend() %>%
  ggplot(offset_labels = -max(SNP.genotype.distances)/25,
             theme = theme_classic()) +
    scale_y_continuous(name = 'SNP distance (manhattan method)',
                                         breaks = function(limits) breaks_pretty()(c(max(limits[1], 0), limits[2]), 5) + max(SNP.genotype.distances)/50,
                                         labels = function(breaks) number(breaks - max(SNP.genotype.distances)/50, big.mark = ','),
                                         limits = c(-max(SNP.genotype.distances)/2, NA),
                                         expand = expansion()) +
    scale_x_continuous(limits = c(0, NA), expand = expansion(add = c(0, .5))) +
    geom_line(data = tibble(x = c(0, 0), y = max(SNP.genotype.distances) * c(1/50, 51/50)),
                        aes(x=x, y=y)) +
  theme(legend.position = "bottom",
            axis.text.x = element_blank(),
            axis.title.x = element_blank(),
            axis.line.x = element_blank(),
            axis.ticks.x = element_blank(),
            axis.line.y = element_blank(),
            axis.title.y = element_text(angle = 90, hjust = .75)
            )

# make legend
gg2 <- dend.format.df %>%
  #arrange(SampleGroup) %>%
  ggplot(aes(x = Sample_ID, y = 1, col = !!sym(color_by), shape = !!sym(shape_by))) +
  geom_point() + 
  scale_color_manual(values = col_map, guide = ifelse(color_by == '_dummy_color', list('none'),
                                        list(guide_legend(direction = 'horizontal',title.position = 'top', ncol = 8, byrow=T)))[[1]] )  +
  scale_shape_manual(values = shape_map, guide = ifelse(shape_by == '_dummy_shape', list('none'),
                                  list(guide_legend(direction = 'horizontal',title.position = 'top', ncol = 8, byrow=T)))[[1]] ) +
  theme(legend.box = "vertical", legend.text = element_text(size = 10), legend.title = element_text(size = 12))

gg <- gg1 + as_ggplot(get_legend(gg2)) + plot_layout(ncol = 1, heights = c(
  5, 1))

subchunkify(gg, 'snp.dendrogram.plot', 8, 10)

```

```{r virtual.karyotype, eval = include.section('virtual.karyotype'), results='asis'}

cat('## Virtual Karyotype\n\n')

if (!is.na(ref_id)) {
  cat(str_glue('The first plot shows only this sample ({sample_name}), the second a side by side comparison  with the reference sample (left: {sample_name}; right: {ref_name}).\n\n'))
}

# Use prepared static data table (genome build specific, UCSC derived) instead of inbuilt RIdeogram
# data(human_karyotype, package="RIdeogram")
genome_data <- gr_info %>% filter(centromer) %>%
  group_by(size) %>%
  reduce_ranges() %>%
  as.data.frame() %>%
  dplyr::rename(Chr = seqnames, CE_start = start, CE_end = end, End = size) %>%
  mutate(Start = 0,
         Chr = str_remove(Chr, 'chr')) %>%
  select(Chr, Start, End, CE_start, CE_end)

sample.calls <- cnv_calls %>%
    dplyr::select(Chr, Start, End, CNV_type, sample_id) %>%
    mutate(Value = ifelse(CNV_type %!in% c('gain', 'loss'), 0, 1),
           Value = ifelse(CNV_type == 'loss', -1, Value),
           Chr = str_remove(Chr, 'chr'))
overlay.tb <- sample.calls %>%
    filter(sample_id == !!sample_id) %>%
    dplyr::select(Chr, Start, End, Value)

#This should not be the case unless there are 0 calls that generated images
if (!dir.exists(file.path(workdir, image_folder))) {
    dir.create(file.path(workdir, image_folder))
}

svg_file1 <- tempfile(fileext = '.svg')
#somehow 'convertSVG' can't deal with absolute paths
png_file1 <- file.path(image_folder, 'virtualKaryotype_ideogram.png')

ideogram(genome_data, overlaid = overlay.tb, colorset1 = c('red', 'grey50', 'green'), output = svg_file1)
convertSVG(svg_file1, file = png_file1)

if (params$out_format == 'html') {
    cat(paste0('<img src="', png_file1, '" width="800" />'))
} else {
  #For reasons tiny-tex/latex executes from the directoy this template is located in (rather than the defined outdir), so relative paths don't work for pdf
  include_graphics(file.path(workdir, png_file1) %>% normalizePath, rel_path = FALSE)
}

if (!is.na(ref_id)) {
  refdata <- sample.calls %>%
        filter(sample_id == ref_id) %>%
        dplyr::select(Chr, Start, End, Value)

  svg_file2 <- tempfile(fileext = '.svg')
  png_file2 <- file.path(image_folder, 'virtualKaryotype_ideogram_vsRef.png')

  ideogram(genome_data, overlaid = overlay.tb, label = refdata, label_type = 'heatmap',
         colorset1 = c('red', 'grey50', 'green'), colorset2 = c('red', 'grey50', 'green'), output = svg_file2)
  convertSVG(svg_file2, file = png_file2)

  if (params$out_format == 'html') {
    cat(paste0('<img src="', png_file2, '" width="800" />'))
  } else {
    include_graphics(file.path(workdir, png_file2) %>% normalizePath, rel_path = FALSE)
  }
}

cat('\n\n')

```
